
```{r}

```
```{r}
chainObject <- rtracklayer::import.chain(file.path(data_dir, 'helpers', 'hg19ToHg38.over.chain'))
```

```{r}
grObject <- GenomicRanges::GRanges(seqnames = cwas_db$chr, ranges = IRanges(start = cwas_db$position, end = cwas_db$position))
results <- as.data.frame(liftOver(grObject, chainObject))
dim(results)
```

- you lose some 422 loci
```{r}
nrow(cwas_db) - nrow(results)
```

- liftover to hg38
```{r}
cvperf_dt_split <- cvperf_dt %>%
    tidyr::separate_wider_delim(col = locus, names = c('chr', 'start', 'end'), delim = stringr::regex(':|-')) %>%
    dplyr::mutate(across(c(start, end), as.numeric)) %>%
    dplyr::select(chr, start, end) %>%
    dplyr::distinct() %>%
    dplyr::mutate(id = 1:nrow(.))

cvperf_granges <- with(cvperf_dt_split, GenomicRanges::GRanges(seqnames = chr, ranges = IRanges(start = start, end = end), id= id))
lfobject <- liftOver(cvperf_granges, chainObject)
# lfobject <- reduce(lfobject, min.gapwidth = 20L)
cvperf_liftover <- as.data.frame(lfobject)
dim(cvperf_liftover)
```

- merge duplicated granges
```{r}
cvperf_liftover <- purrr::map2(list(min, max), list('start', 'end'), ~ 
      cvperf_liftover %>%
         group_by(id) %>%
         summarise_at(.y, .x)) %>%
         purrr::reduce(inner_join, by = 'id') %>%
         dplyr::rename(start.hg38 = start, end.hg38 = end)

cvperf_liftover[1:5, ]
```

```{r}
cwas_bed_mappings <- cvperf_dt_split %>% dplyr::rename(start.hg19 = start, end.hg19 = end) %>%
    dplyr::inner_join(cvperf_liftover, by = 'id') %>%
    dplyr::mutate(cwas_locus.hg19 = paste0(chr, ':', start.hg19, '-', end.hg19),
        cwas_locus.hg38 = paste0(chr, '_', start.hg38, '_', end.hg38))
```

```{r}
hg38bedfiles <- hg38_bed_files %>%
    dplyr::mutate(hg38_id = paste0(chr, '_', hg38_start, '_', hg38_end))
```

```{r}
which(!cwas_bed_mappings$cwas_locus.hg38 %in% hg38bedfiles$hg38_id); which(!hg38bedfiles$hg38_id %in% cwas_bed_mappings$cwas_locus.hg38)
```

```{r}
hg38bedfiles[186, ] ; cwas_bed_mappings %>% dplyr::filter(cwas_locus.hg19 == 'chr1:145543400-145544050')
```


```{r}
length(cvperf_liftover$id |> unique()) - length(cvperf_dt_split$id)
```

```{r}
cvperf_liftover %>% 
  group_by(id) %>% 
  filter(n()>1)
```

```{r}
cvperf_liftover %>%
    dplyr::group_by(id) %>%
    filter(start == min(start), end == max(end)) %>%
    filter(n()>1)
```

- split into models and save 
```{r}
dt <- cvperf_dt %>% 
    tidyr::pivot_longer(cols = !c(locus, measure, transcription_factor, n_snps_in_window, n.snps.in.model), values_to = 'score', names_to=c('model')) %>% 
    tidyr::pivot_wider(id_cols = c(locus, model, transcription_factor, n_snps_in_window, n.snps.in.model), names_from = 'measure', values_from = 'score') %>%
    dplyr::mutate(locus = gsub(':|-', '_', locus)) %>%
    dplyr::full_join(loci_mappings, by = c('locus' = 'hg19_id')) %>%
    dplyr::rename(hg19_id = locus, pred.perf.R2 = rsq, pred.perf.pval = pval)

dt %>% dplyr::group_by(model) %>%
    group_walk(~ data.table::fwrite(.x, file=glue('{files_dir}/{transcription_factor}_baca_cwas_{tdate}.{.y$model}.cv_performance.txt.gz'), col.names=T, row.names=F, quote=F, compress='gzip',sep = '\t'), .keep = TRUE)
```


```{r}
cwas_db <- cwas_db %>%
    dplyr::mutate(varID = gsub(":|-", '_', locus))
extra_dt <- extra_dt %>%
    dplyr::mutate(varID = gsub(":|-", '_', locus))
```

```{r}
bedmappings <- glue('{base_dir}/mappings/baca_cwas_loci_hg38.bed')
snpmappings <- glue('{base_dir}/mappings/hg38_snps.bed')
```

```{r}
hg38_snp_files <- data.table::fread(snpmappings, col.names=c('chr', 'hg38_start', 'hg38_end', 'snp_id', 'hg19_locus'))
hg38_snp_files$chr <- as.numeric(gsub(pattern='chr', replacement='', x=hg38_snp_files$chr))
hg38_snp_files[1:5, ]
```

```{r}
tt <- dplyr::left_join(hg38_cwas_db, hg38_snp_files, by=c('locus' = 'hg38_locus', 'snp_id'='snp_id', 'chr' = 'chr'))
tt$position <- tt$hg38_start
tt$hg38_start <- tt$hg38_end <- NULL
tt[1:5, ] ; dim(tt)
```

```{r}
mappings <- data.table::fread('/project2/haky/temi/projects/TFXcan/baca_cwas/mappings/baca_cwas_loci_hg38.bed', col.names=c('chr', 'hg38_start', 'hg38_end', 'hg19_id'))

mappings <- mappings %>%
    tidyr::unite('hg38_id', chr:hg38_end, sep = '_', remove = T) %>%
    dplyr::mutate(hg19_id = gsub(':|-', '_', hg19_id))

data.table::fwrite(mappings, file = '/project2/haky/temi/projects/TFXcan/baca_cwas/mappings/cwas_loci_mappings.txt', sep = '\t', col.names = T, row.names = F, quote = F)
```

loci_mappings <- data.table::fread('/project2/haky/temi/projects/TFXcan/baca_cwas/mappings/cwas_loci_mappings.txt')
data.table::fwrite(extra_dt, file=glue('{files_dir}/{transcription_factor}_baca_cwas_extras_{tdate}.hg19.txt.gz'), row.names=F, quote=F, compress='gzip', sep='\t')
```

#### Create db (in hg19)
Baca has 6 models/weights

```{r}
baca_models <- c('lasso', 'lasso.as', 'lasso.plasma', 'top1.as', 'top1.qtl', 'top1')
db_folder <- glue('{base_dir}/db_folder')
if(!dir.exists(db_folder)){dir.create(db_folder)}

db_folder_chr <- glue('{base_dir}/db_folder_chr')
if(!dir.exists(db_folder_chr)){dir.create(db_folder_chr)}
```

```{r}
baca_extra <- data.table::fread(glue('{files_dir}/{transcription_factor}_baca_cwas_extras_{tdate}.hg19.txt.gz'))
baca_extra <- baca_extra %>% 
    dplyr::mutate(gene = gsub(":|-", '_', locus)) %>%
    dplyr::select(-c(varID)) %>%
    dplyr::rename(genename=transcription_factor)
baca_extra$pred.perf.qval <- NA
baca_extra[1:5, ]
```

Predict_db format

```{r}

weights_dt <- data.table::fread(glue('{files_dir}/{transcription_factor}_baca_cwas_weights_{tdate}.hg19.txt.gz'))

baca_weights_list <- purrr::map(.x=baca_models, function(each_m){
    model_weights <- weights_dt %>% 
        dplyr::mutate(varID = paste0(chr, '_', position, '_', a1, '_', a2, sep=''),
            chr_varID = paste0('chr', chr, '_', position, '_', a1, '_', a2, sep=''),
            gene = gsub(":|-", '_', locus)) %>%
        dplyr::select(gene, rsid=snp_id, varID, chr_varID, ref_allele=a1, eff_allele=a2, weight=as.symbol(each_m)) %>% as.data.frame()

    # read in the extra
    extras_dt <- data.table::fread(glue('{files_dir}/{transcription_factor}_baca_cwas_{tdate}.{each_m}.cv_performance.txt.gz'))

    # prepare for both hg19 and hg38

    ## hg19 
    each_db <- DBI::dbConnect(RSQLite::SQLite(), glue('{db_folder}/baca_cwas_{each_m}.hg19.db'))
    dbWriteTable(each_db, "extra", extras_dt %>% dplyr::rename(gene = hg19_id), overwrite=T)
    dbWriteTable(each_db, "weights", model_weights, overwrite=T)
    dbDisconnect(each_db)

    ## hg38
    edt <- extras_dt %>%
        dplyr::select(hg19_id, hg38_id)
    model_wgts <- dplyr::full_join(model_weights, edt, by = c('gene' = 'hg19_id')) %>%
        dplyr::select(-any_of(c(gene, hg19_id))) %>%
        dplyr::rename(gene = hg38_id)


    each_db <- DBI::dbConnect(RSQLite::SQLite(), glue('{db_folder}/baca_cwas_{each_m}.hg19.db'))
    dbWriteTable(each_db, "extra", baca_extra, overwrite=T)
    dbWriteTable(each_db, "weights", model_weights, overwrite=T)
    dbDisconnect(each_db)

    # model_weights <- baca_weights %>% 
    #     dplyr::select(gene=locus, rsid=snp_id, varID=chr_varIDs, chr_varIDs=chr_varIDs, ref_allele=a1, eff_allele=a2, weight=as.symbol(each_m)) %>% as.data.frame()

    # each_db <- DBI::dbConnect(RSQLite::SQLite(), glue('{db_folder_chr}/baca_cwas_{each_m}.db'))
    # dbWriteTable(each_db, "extra", baca_extra, overwrite=T)
    # dbWriteTable(each_db, "weights", model_weights, overwrite=T)
    # dbDisconnect(each_db)

    return(0)
})

# names(baca_weights_list) <- baca_models
```

### Read in the cv performance

```{r}
out <- purrr::map(.x=seq_along(ar_files_locus), .f=function(i){
    locus <- ar_files_locus[i]
    #print(file.exists(glue('{output_dir}/{transcription_factor}/{locus}.wgt.RDat')))
    rdt <- new.env(parent = emptyenv())
    load(glue('{output_dir}/{transcription_factor}/{locus}.wgt.RDat'), envir=rdt)

    cv_perf <- rbind(rdt$cv.performance['pval', ], rdt$cv.performance['rsq', ]) %>%
        as.data.frame() %>%
        dplyr::mutate(measure = c('pval', 'rsq'), snp_id = locus) %>%
        dplyr::relocate(snp_id, measure)

    return(cv_perf)
}, .progress=T)

cvperf_dt <- do.call('rbind', out)
cvperf_dt <- cvperf_dt %>% dplyr::rename(locus = snp_id)

dim(cvperf_dt) ; cvperf_dt[1:5, ]
```

```{r}
dt <- dplyr::left_join(cvperf_dt, hg38_bed_files, by=c('locus' = 'hg19_id')) %>%
    dplyr::mutate(locus = hg38_id) %>%
    dplyr::rename(chrom = chr) %>%
    dplyr::relocate(chrom, locus, measure) %>%
    dplyr::select(-hg38_id) %>%
    dplyr::filter(!is.na(locus))
```

```{r}
sum(is.na(dt$locus))
```

```{r}
dt %>% 
    tidyr::pivot_longer(cols = !c(chrom, locus, measure), values_to = 'score', names_to=c('model')) %>% 
    tidyr::pivot_wider(id_cols =c(locus, model),  names_from = 'measure', values_from = 'score') %>%
    dplyr::group_by(model) %>%
    group_walk(~ data.table::fwrite(.x, file=glue('{files_dir}/{transcription_factor}_baca_cwas_hg38_{tdate}.{.y$model}.cv_performance.txt.gz'), col.names=T, row.names=F, quote=F, compress='gzip',sep = '\t'), .keep = TRUE)
```

```{r}
dtgrp[[1]] %>% dplyr::group_by(locus, model, measure) %>%
  dplyr::summarise(n = dplyr::n(), .groups = "drop") %>%
  dplyr::filter(n > 1L) 
```

```{r}
dt <- dplyr::left_join(cvperf_dt, hg38_bed_files, by=c('locus' = 'hg19_id')) %>% 
    dplyr::mutate(locus = hg38_id) %>%
    dplyr::rename(chrom = chr) %>%
    dplyr::relocate(chrom, locus, measure) %>%
    dplyr::select(-hg38_id)
data.table::fwrite(dt, file=glue('{files_dir}/{transcription_factor}_baca_cwas_hg38_{tdate}.cv_performance.txt.gz'), col.names=T, row.names=F, quote=F, compress='gzip',sep = '\t')
```

```{r}
dt <- data.table::fread(glue('{files_dir}/{transcription_factor}_baca_cwas_hg38_{tdate}.cv_performance.txt.gz'))
```

```{r}
pval_mat <- dt %>%
    dplyr::filter(measure == 'pval') %>%
    dplyr::select(-c(chrom, locus, measure)) %>%
    as.matrix() 

rsq_mat <- dt %>%
    dplyr::filter(measure == 'rsq') %>%
    dplyr::select(-c(chrom, locus, measure)) %>%
    as.matrix() 
```

```{r}
# Function to add histograms
panel.hist <- function(x, ...) {
    # usr <- par("usr")
    # on.exit(par(usr))
    # par(usr = c(usr[1:2], 0, 1.5))
    par(new = TRUE)
    his <- hist(x, plot = T, main = '')
    # breaks <- his$breaks
    # nB <- length(breaks)
    # y <- his$counts
    # y <- y/max(y)
    # rect(breaks[-nB], 0, breaks[-1], y, col = rgb(0, 1, 1, alpha = 0.5), ...)
    # lines(density(x), col = 2, lwd = 2) # Uncomment to add density lines
}

# panel.qqplot <- function(x, ...) {
#     usr <- par("usr")
#     on.exit(par(usr))
#     par(usr = c(usr[1:2], 0, 1.5))
#     qq_generic(data_points = x, distribution = 'uniform')
#     # lines(density(x), col = 2, lwd = 2) # Uncomment to add density lines
# }


panel.cor <- function(x,y, ...){
    par(new = TRUE)
    plot(x, y)
    abline(a=0, b=1, col='red')
    # legend('bottomright', legend=c('y', 'x'), col=c("black", "blue"), pch=c(1,3), bg = adjustcolor('grey', 0.2))
}

panel.qqplot <- function(x,y, ...){
    par(new = TRUE)
    qqunif.compare(x, y)
    legend('bottomright', legend=c('y', 'x'), col=c("black", "blue"), pch=c(1,3), bg = adjustcolor('grey', 0.2))
}
```



```{r}
qqunif(pval_mat[, 'lasso'], col = 'orange', pch = '.', cex = 3, bty = 'n', show=F)
qqpoints(pval_mat[, 'lasso.as'], col = 'brown', pch = '.', cex = 3)
qqpoints(pval_mat[, 'top1'], col = 'red', pch = '.', cex = 3)
qqpoints(pval_mat[, 'lasso.plasma'], col = 'blue', pch = '.', cex = 3)
qqpoints(pval_mat[, 'top1.as'], col = 'green', pch = '.', cex = 3)
qqpoints(pval_mat[, 'top1.qtl'], col = 'black', pch = '.', cex = 3)
legend(x=0, y =20, legend = c('lasso', 'lasso.as', 'top1', 'lasso.plasma', 'top1.as', 'top1.qtl'), pch = '.', pt.cex=3, col=c('orange', 'brown', 'red', 'blue', 'green', 'black'), bty='n')

```

```{r}
# pdf(glue('{files_dir}/plt.pdf'), width=21, height=21)
# pairs(pval_mat, lower.panel = panel.qqplot, diag.panel = panel.hist, upper.panel = NULL, gap=3)
# dev.off()

pdf(glue('{files_dir}/rsq_plt.pdf'), width=21, height=21)
pairs(rsq_mat, lower.panel = panel.cor, diag.panel = panel.hist, upper.panel = NULL, gap=3)
dev.off()

```

```{r}
qqplot(pval_mat[1:100, 1], pval_mat[1:100, 2])
```

if you have a lifted over bed file, you can continue
Read in new bed files and match 
```{r}
hg38_bed_files <- data.table::fread(bedmappings, col.names=c('chr', 'hg38_start', 'hg38_end', 'hg19_id'))
hg38_bed_files <- hg38_bed_files %>% 
    dplyr::mutate(hg38_id=paste(paste(chr, hg38_start, sep=':'), hg38_end, sep='-')) %>%
    dplyr::select(chr, hg19_id, hg38_id)
hg38_bed_files$chr <- as.numeric(gsub(pattern='chr', replacement='', x=hg38_bed_files$chr))
hg38_bed_files[1:5, ]
```

```{r}
hg38_cwas_db <- dplyr::left_join(cwas_db, hg38_bed_files, by=c('locus' = 'hg19_id', 'chr'='chr'))
hg38_cwas_db$locus <- hg38_cwas_db$hg38_id
hg38_cwas_db$hg38_id <- NULL
hg38_cwas_db[1:5, ] ; dim(hg38_cwas_db)
```

Match the `extras` too

```{r}
hg38_cwas_extra <- dplyr::left_join(extra_dt, hg38_bed_files, by=c('locus' = 'hg19_id'))
hg38_cwas_extra$locus <- hg38_cwas_extra$hg38_id
hg38_cwas_extra$hg38_id <- NULL
hg38_cwas_extra[1:5, ] ; dim(hg38_cwas_extra)
```

Merge with the hg38 snps
```{r}
hg38_snp_files <- data.table::fread(snpmappings, col.names=c('chr', 'hg38_start', 'hg38_end', 'snp_id', 'hg38_locus'))
hg38_snp_files$chr <- as.numeric(gsub(pattern='chr', replacement='', x=hg38_snp_files$chr))
hg38_snp_files[1:5, ]
```

```{r}
tt <- dplyr::left_join(hg38_cwas_db, hg38_snp_files, by=c('locus' = 'hg38_locus', 'snp_id'='snp_id', 'chr' = 'chr'))
tt$position <- tt$hg38_start
tt$hg38_start <- tt$hg38_end <- NULL
tt[1:5, ] ; dim(tt)
```

```{r}
# write out the snps to a file
data.table::fwrite(tt, file=glue('{files_dir}/{transcription_factor}_baca_cwas_weights_hg38_{tdate}.txt.gz'), col.names=T, row.names=F, quote=F, compress='gzip',sep = '\t')

data.table::fwrite(hg38_cwas_extra, file=glue('{files_dir}/{transcription_factor}_baca_cwas_extra_hg38_{tdate}.txt.gz'), col.names=T, row.names=F, quote=F, compress='gzip',sep = '\t')
```


## Create databases

- for the weights as is
Here I will write out two folders: one with `1...` and `chr1...`
```{r}
library(RSQLite)
library(dbplyr)
library(DBI)
```

```{r}
db_folder <- glue('{base_dir}/db_folder')
if(!dir.exists(db_folder)){dir.create(db_folder)}

db_folder_chr <- glue('{base_dir}/db_folder_chr')
if(!dir.exists(db_folder_chr)){dir.create(db_folder_chr)}
```


```{r}
# write out the snps to a file
baca_weights <- data.table::fread(glue('{files_dir}/{transcription_factor}_baca_cwas_weights_hg38_{tdate}.txt.gz'))
baca_weights$varIDs <- with(baca_weights, paste0(chr, '_', position, '_', a1, '_', a2, sep=''))
baca_weights$chr_varIDs <- with(baca_weights, paste0('chr', chr, '_', position, '_', a1, '_', a2, sep=''))
baca_weights[1:5, ]
```

Baca has 6 models/weights

```{r}
baca_models <- c('lasso', 'lasso.as', 'lasso.plasma', 'top1.as', 'top1.qtl', 'top1')
```

```{r}
baca_extra <- data.table::fread(glue('{files_dir}/{transcription_factor}_baca_cwas_extra_hg38_{tdate}.txt.gz'))
baca_extra <- baca_extra %>% dplyr::rename(gene=locus, genename=transcription_factor)
baca_extra$pred.perf.qval <- NA
baca_extra[1:5, ]
```

Predict_db format

```{r}
baca_weights_list <- purrr::map(.x=baca_models, function(each_m){
    model_weights <- baca_weights %>% 
        dplyr::select(gene=locus, rsid=snp_id, varID=varIDs, chr_varID=chr_varIDs, ref_allele=a1, eff_allele=a2, weight=as.symbol(each_m)) %>% as.data.frame()

    each_db <- DBI::dbConnect(RSQLite::SQLite(), glue('{db_folder}/baca_cwas_{each_m}.db'))
    dbWriteTable(each_db, "extra", baca_extra, overwrite=T)
    dbWriteTable(each_db, "weights", model_weights, overwrite=T)
    dbDisconnect(each_db)

    # model_weights <- baca_weights %>% 
    #     dplyr::select(gene=locus, rsid=snp_id, varID=chr_varIDs, chr_varIDs=chr_varIDs, ref_allele=a1, eff_allele=a2, weight=as.symbol(each_m)) %>% as.data.frame()

    # each_db <- DBI::dbConnect(RSQLite::SQLite(), glue('{db_folder_chr}/baca_cwas_{each_m}.db'))
    # dbWriteTable(each_db, "extra", baca_extra, overwrite=T)
    # dbWriteTable(each_db, "weights", model_weights, overwrite=T)
    # dbDisconnect(each_db)

    return(0)
})

# names(baca_weights_list) <- baca_models
```




Now you can look at one of them...
```{r}
lasso_db <- DBI::dbConnect(RSQLite::SQLite(), glue('{db_folder}/baca_cwas_lasso.db'))
extra_dt <- tbl(lasso_db, 'extra') %>% as.data.frame()
weights_dt <- tbl(lasso_db, 'weights') %>% as.data.frame()

dbDisconnect(lasso_db)
```

```{r}
extra_dt |> head() ; weights_dt |> head()
```



```{r}
lasso_db <- DBI::dbConnect(RSQLite::SQLite(), glue('{db_folder_chr}/baca_cwas_lasso.db'))
extra_dt <- tbl(lasso_db, 'extra') %>% as.data.frame()
weights_dt <- tbl(lasso_db, 'weights') %>% as.data.frame()

dbDisconnect(lasso_db)
```

```{r}
extra_dt |> head() ; weights_dt |> head()
```



```{r}
zcat in.vcf.gz | vcftools_0.1.9/bin/vcf-annotate --fill-type | grep -oP "TYPE=\w+" | sort | uniq -c
```



```{r}
md <- readRDS('/project/haky/users/temi/Enpact-figures/data/aggByCollect_AR_Prostate.logistic.rds')
```

```{r}
a <- data.table::fread('/project/haky/users/temi/Enpact-figures/data/aggByCollect_AR_Prostate.logistic.test_eval.txt.gz')

b <- data.table::fread('/project2/haky/temi/projects/TFPred-snakemake/output/models_eval/cistrome_AR_Prostate_2024-04-23/aggByCollect_AR_Prostate.logistic.test_eval.txt.gz')

c <- data.table::fread('/project2/haky/temi/projects/TFPred-snakemake/output/models_eval/cistrome_AR_Prostate_2024-04-23/aggByCollect_AR_Prostate.logistic.train_eval.txt.gz')
```


```{r}
b %>%
    dplyr::select(locus, binding_class, binding_count) %>%
    dplyr::mutate(split = 'test') %>%
    data.table::fwrite(file=file.path(data_dir, 'data/enpact/files/', 'AR_Prostate.test_arbs.txt.gz'), col.names=T, row.names=F, quote=F, compress='gzip', sep='\t')

c %>%
    dplyr::select(locus, binding_class, binding_count) %>%
    dplyr::mutate(split = 'train') %>%
    data.table::fwrite(file=file.path(data_dir, 'data/enpact/files/', 'AR_Prostate.train_arbs.txt.gz'), col.names=T, row.names=F, quote=F, compress='gzip', sep='\t')
```


```{r}
bind_rows(
    dplyr::select(b, locus, binding_class, binding_count) %>%
    dplyr::mutate(split = 'test'),
    dplyr::select(c, locus, binding_class, binding_count) %>%
    dplyr::mutate(split = 'train')
) %>%
    data.table::fwrite(file=file.path(data_dir, 'data/enpact/files/', 'AR_Prostate.arbs_groundtruth.txt.gz'), col.names=T, row.names=F, quote=F, compress='gzip', sep='\t')
```


```{r}
data.table::fread(file.path(data_dir, 'data/enpact/files/', 'AR_Prostate.arbs_groundtruth.txt.gz')) %>%
    dplyr::select(locus) %>%
    data.table::fwrite(file=file.path(data_dir, 'data/enpact/files/', 'AR_Prostate.arbs_predict.txt'), col.names=F, row.names=F, quote=F, sep='\t')
```


```{r}
input <- file.path(data_dir, 'data/enpact/files', 'AR_Prostate.arbs_predict.txt')
reference_epigenome_dir <- '/project2/haky/Data/enformer-reference-epigenome'
output <- file.path(data_dir, 'data/enpact/files', 'AR_Prostate.arbs_epigenome.csv.gz')
exec_py <- '/project/haky/users/temi/projects/TFPred-snakemake/workflow/src/aggregate_epigenomes.py'

cmd <- glue("sbatch {data_dir}/src/infer.sbatch {exec_py} {input} {reference_epigenome_dir} {output} --use_multiprocessing")

cmd
```


```{r}
system(cmd) ; system('squeue -u temi')
```


```{r}
dtepi <- data.table::fread('/project/haky/users/temi/projects/Enpact/data/enpact/files/cistrome_aggByCollect_AR_Prostate.csv.gz')
gt <- data.table::fread(file.path(data_dir, 'data/enpact/files', 'AR_Prostate.arbs_groundtruth.txt.gz'))
gt_train <- gt %>% dplyr::filter(split == 'train')
gt_test <- gt %>% dplyr::filter(split == 'test')
```


```{r}
dt_train <- dtepi %>% dplyr::filter(id %in% gt_train$locus)
dt_gt <- gt_train %>% dplyr::select(-split)
dt_train <- inner_join(dt_gt, dt_train, by=c('locus' = 'id'))

dim(dt_train); dt_train[1:5, 1:10]

data.table::fwrite(dt_train, '/project/haky/users/temi/projects/Enpact/data/enpact/training/AR_Prostate.train_epigenome.csv.gz', col.names=T, row.names=F, quote=F, compress='gzip', sep='\t')
```


```{r}
dt_test <- dtepi %>% dplyr::filter(id %in% gt_test$locus)
dt_gt <- gt_test %>% dplyr::select(-split)
dt_test <- inner_join(dt_gt, dt_test, by=c('locus' = 'id'))
dim(dt_test); dt_test[1:5, 1:10]
data.table::fwrite(dt_test, '/project/haky/users/temi/projects/Enpact/data/enpact/training/AR_Prostate.test_epigenome.csv.gz', col.names=T, row.names=F, quote=F, compress='gzip', sep='\t')
```

FOXA1 

```{r}
a <- data.table::fread('/project/haky/users/temi/Enpact-figures/data/aggByCollect_FOXA1_Prostate.logistic.test_eval.txt.gz')

b <- data.table::fread('/project2/haky/temi/projects/TFPred-snakemake/output/models_eval/cistrome_FOXA1_Prostate_2024-04-23/aggByCollect_FOXA1_Prostate.logistic.test_eval.txt.gz')

c <- data.table::fread('/project2/haky/temi/projects/TFPred-snakemake/output/models_eval/cistrome_FOXA1_Prostate_2024-04-23/aggByCollect_FOXA1_Prostate.logistic.train_eval.txt.gz')
```


```{r}
b %>%
    dplyr::select(locus, binding_class, binding_count) %>%
    dplyr::mutate(split = 'test') %>%
    data.table::fwrite(file=file.path(data_dir, 'data/enpact/files/', 'FOXA1_Prostate.test_arbs.txt.gz'), col.names=T, row.names=F, quote=F, compress='gzip', sep='\t')

c %>%
    dplyr::select(locus, binding_class, binding_count) %>%
    dplyr::mutate(split = 'train') %>%
    data.table::fwrite(file=file.path(data_dir, 'data/enpact/files/', 'FOXA1_Prostate.train_arbs.txt.gz'), col.names=T, row.names=F, quote=F, compress='gzip', sep='\t')
```


```{r}
bind_rows(
    dplyr::select(b, locus, binding_class, binding_count) %>%
    dplyr::mutate(split = 'test'),
    dplyr::select(c, locus, binding_class, binding_count) %>%
    dplyr::mutate(split = 'train')
) %>%
    data.table::fwrite(file=file.path(data_dir, 'data/enpact/files/', 'FOXA1_Prostate.arbs_groundtruth.txt.gz'), col.names=T, row.names=F, quote=F, compress='gzip', sep='\t')
```


```{r}
dtepi <- data.table::fread('/project/haky/users/temi/projects/Enpact/data/enpact/files/cistrome_aggByCollect_FOXA1_Prostate.csv.gz')
gt <- data.table::fread(file.path(data_dir, 'data/enpact/files', 'FOXA1_Prostate.arbs_groundtruth.txt.gz'))
gt_train <- gt %>% dplyr::filter(split == 'train')
gt_test <- gt %>% dplyr::filter(split == 'test')
```


```{r}
dt_train <- dtepi %>% dplyr::filter(id %in% gt_train$locus)
dt_gt <- gt_train %>% dplyr::select(-split)
dt_train <- inner_join(dt_gt, dt_train, by=c('locus' = 'id'))

dim(dt_train); dt_train[1:5, 1:10]

data.table::fwrite(dt_train, '/project/haky/users/temi/projects/Enpact/data/enpact/training/FOXA1_Prostate.train_epigenome.csv.gz', col.names=T, row.names=F, quote=F, compress='gzip', sep='\t')
```


```{r}
dt_test <- dtepi %>% dplyr::filter(id %in% gt_test$locus)
dt_gt <- gt_test %>% dplyr::select(-split)
dt_test <- inner_join(dt_gt, dt_test, by=c('locus' = 'id'))
dim(dt_test); dt_test[1:5, 1:10]
data.table::fwrite(dt_test, '/project/haky/users/temi/projects/Enpact/data/enpact/training/FOXA1_Prostate.test_epigenome.csv.gz', col.names=T, row.names=F, quote=F, compress='gzip', sep='\t')
```

```{r}
for(tf in c('HOXB13', 'GATA2')){
    aa <- data.table::fread(glue('/project/haky/users/temi/Enpact-figures/data/aggByCollect_{tf}_Prostate.logistic.test_eval.txt.gz'))

    bb <- data.table::fread(glue('/project2/haky/temi/projects/TFPred-snakemake/output/models_eval/cistrome_{tf}_Prostate_2024-04-23/aggByCollect_{tf}_Prostate.logistic.test_eval.txt.gz'))

    cc <- data.table::fread(glue('/project2/haky/temi/projects/TFPred-snakemake/output/models_eval/cistrome_{tf}_Prostate_2024-04-23/aggByCollect_{tf}_Prostate.logistic.train_eval.txt.gz'))

    bb %>%
    dplyr::select(locus, binding_class, binding_count) %>%
    dplyr::mutate(split = 'test') %>%
    data.table::fwrite(file=file.path(data_dir, 'data/enpact/files/', glue('{tf}_Prostate.test_arbs.txt.gz')), col.names=T, row.names=F, quote=F, compress='gzip', sep='\t')

    cc %>%
        dplyr::select(locus, binding_class, binding_count) %>%
        dplyr::mutate(split = 'train') %>%
        data.table::fwrite(file=file.path(data_dir, 'data/enpact/files/', glue('{tf}_Prostate.train_arbs.txt.gz')), col.names=T, row.names=F, quote=F, compress='gzip', sep='\t')

    bind_rows(
        dplyr::select(bb, locus, binding_class, binding_count) %>%
        dplyr::mutate(split = 'test'),
        dplyr::select(cc, locus, binding_class, binding_count) %>%
        dplyr::mutate(split = 'train')
    ) %>%
        data.table::fwrite(file=file.path(data_dir, 'data/enpact/files/', glue('{tf}_Prostate.arbs_groundtruth.txt.gz')), col.names=T, row.names=F, quote=F, compress='gzip', sep='\t')

    dtepi <- data.table::fread(glue('/project/haky/users/temi/projects/Enpact/data/enpact/files/cistrome_aggByCollect_{tf}_Prostate.csv.gz'))
    gt <- data.table::fread(file.path(data_dir, 'data/enpact/files', glue('{tf}_Prostate.arbs_groundtruth.txt.gz')))
    gt_train <- gt %>% dplyr::filter(split == 'train')
    gt_test <- gt %>% dplyr::filter(split == 'test')

    dt_train <- dtepi %>% dplyr::filter(id %in% gt_train$locus)
    dt_gt <- gt_train %>% dplyr::select(-split)
    dt_train <- inner_join(dt_gt, dt_train, by=c('locus' = 'id'))

    dim(dt_train); dt_train[1:5, 1:10]

    data.table::fwrite(dt_train, glue('/project/haky/users/temi/projects/Enpact/data/enpact/training/{tf}_Prostate.train_epigenome.csv.gz'), col.names=T, row.names=F, quote=F, compress='gzip', sep='\t')

    dt_test <- dtepi %>% dplyr::filter(id %in% gt_test$locus)
    dt_gt <- gt_test %>% dplyr::select(-split)
    dt_test <- inner_join(dt_gt, dt_test, by=c('locus' = 'id'))
    dim(dt_test); dt_test[1:5, 1:10]
    data.table::fwrite(dt_test, glue('/project/haky/users/temi/projects/Enpact/data/enpact/training/{tf}_Prostate.test_epigenome.csv.gz'), col.names=T, row.names=F, quote=F, compress='gzip', sep='\t')
}
```

