<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.269">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Temi">
<meta name="dcterms.date" content="2022-12-14">

<title>IMPACT vs.&nbsp;PMIntersect (Peak Motif Intersect) approaches to defining bound and unbound regions</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="impact_line_by_line_files/libs/clipboard/clipboard.min.js"></script>
<script src="impact_line_by_line_files/libs/quarto-html/quarto.js"></script>
<script src="impact_line_by_line_files/libs/quarto-html/popper.min.js"></script>
<script src="impact_line_by_line_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="impact_line_by_line_files/libs/quarto-html/anchor.min.js"></script>
<link href="impact_line_by_line_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="impact_line_by_line_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="impact_line_by_line_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="impact_line_by_line_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="impact_line_by_line_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">IMPACT vs.&nbsp;PMIntersect (Peak Motif Intersect) approaches to defining bound and unbound regions</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Temi </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">December 14, 2022</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glue)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rjson)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: stats4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    anyDuplicated, append, as.data.frame, basename, cbind, colnames,
    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
    union, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:data.table':

    first, second</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: IRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'IRanges'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:data.table':

    shift</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:glue':

    trim</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(parallel)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching packages
───────────────────────────────────────
tidyverse 1.3.2 ──</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>✔ ggplot2 3.3.6     ✔ purrr   0.3.4
✔ tibble  3.1.8     ✔ dplyr   1.0.9
✔ tidyr   1.2.0     ✔ stringr 1.4.0
✔ readr   2.1.2     ✔ forcats 0.5.1
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::between()    masks data.table::between()
✖ dplyr::collapse()   masks IRanges::collapse()
✖ dplyr::combine()    masks BiocGenerics::combine()
✖ dplyr::desc()       masks IRanges::desc()
✖ tidyr::expand()     masks S4Vectors::expand()
✖ dplyr::filter()     masks stats::filter()
✖ dplyr::first()      masks S4Vectors::first(), data.table::first()
✖ dplyr::lag()        masks stats::lag()
✖ dplyr::last()       masks data.table::last()
✖ ggplot2::Position() masks BiocGenerics::Position(), base::Position()
✖ purrr::reduce()     masks GenomicRanges::reduce(), IRanges::reduce()
✖ dplyr::rename()     masks S4Vectors::rename()
✖ dplyr::slice()      masks IRanges::slice()
✖ purrr::transpose()  masks data.table::transpose()</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>project_dir <span class="ot">&lt;-</span> <span class="st">'/lus/theta-fs0/projects/covid-ct/imlab/users/temi/projects/TFXcan/impact_pipeline'</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="co">#setwd(project_dir)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>output_dir <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'{project_dir}/explore_impact'</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>id <span class="ot">&lt;-</span> <span class="st">'kawakami'</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>TF <span class="ot">&lt;-</span> <span class="st">'FOXA1'</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>cell_line <span class="ot">&lt;-</span> <span class="st">'LuCaP'</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>homer_dir <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'~/miniconda3/envs/homer-env/share/homer'</span>)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>common_dir <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'{project_dir}/processed_data/impact_files/common_files/{TF}'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Directories and e.t.c.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>chip_file <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'{project_dir}/processed_data/impact_files/kawakami-human/bedtools-intersected-regions/FOXA1_intersected_files.bed'</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>chip_dt <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(chip_file, <span class="at">select=</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>))</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>chip_dt[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   chrom start   end num        list
1:  chr1  9938  9948   1          10
2:  chr1  9948  9975   2       10,12
3:  chr1  9975 10116   3    10,12,51
4:  chr1 10116 10284   4 10,12,15,51
5:  chr1 10284 10296   3    10,12,15</code></pre>
</div>
</div>
<p>Select 200 peaks to test with (These are randomly selected peaks)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2022</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>chip_dt <span class="ot">&lt;-</span> chip_dt[<span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(chip_dt), <span class="dv">200</span>, <span class="at">replace=</span>F), ]</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(chip_dt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 200</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>chip_granges <span class="ot">&lt;-</span> <span class="fu">with</span>(chip_dt, GenomicRanges<span class="sc">::</span><span class="fu">GRanges</span>(chrom, <span class="fu">IRanges</span>(start, end), <span class="at">strand=</span><span class="st">'*'</span>))</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>chip_granges</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 200 ranges and 0 metadata columns:
        seqnames              ranges strand
           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;
    [1]     chrX 103122567-103122585      *
    [2]     chr6   37842908-37842911      *
    [3]     chr5 142305491-142305568      *
    [4]     chr1 164453204-164453206      *
    [5]     chr6   42922199-42922204      *
    ...      ...                 ...    ...
  [196]    chr18   74902999-74903006      *
  [197]     chr2 144059425-144059465      *
  [198]    chr13 114548061-114548066      *
  [199]     chr6   44526815-44526818      *
  [200]     chr6 144736546-144736560      *
  -------
  seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Next, make the granges appropriate for impact</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>chip_bed <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">as.character</span>(<span class="fu">seqnames</span>(chip_granges)), <span class="fu">start</span>(chip_granges), <span class="fu">end</span>(chip_granges), <span class="fu">paste0</span>(<span class="st">"peak"</span>, <span class="fu">seq</span>(<span class="dv">1</span>,<span class="fu">length</span>(chip_granges),<span class="dv">1</span>)), <span class="dv">0</span>, <span class="st">"+"</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(chip_bed, <span class="fu">glue</span>(<span class="st">"{output_dir}/{id}_{TF}_{cell_line}_train_df.txt"</span>), <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">quote =</span> F, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>chip_bed[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1]   [,2]        [,3]        [,4]    [,5] [,6]
[1,] "chrX" "103122567" "103122585" "peak1" "0"  "+" 
[2,] "chr6" "37842908"  "37842911"  "peak2" "0"  "+" 
[3,] "chr5" "142305491" "142305568" "peak3" "0"  "+" 
[4,] "chr1" "164453204" "164453206" "peak4" "0"  "+" 
[5,] "chr6" "42922199"  "42922204"  "peak5" "0"  "+" </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>cmd <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'perl {homer_dir}/bin/findMotifsGenome.pl {output_dir}/{id}_{TF}_{cell_line}_train_df.txt {homer_dir}/data/genomes/hg19 {output_dir}/homer_output/ -size given -find {common_dir}/{TF}_{cell_line}.motif &gt; {output_dir}/{id}_{TF}_{cell_line}_findinstances.txt'</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(cmd)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="positive-regions-are-defined-this-way" class="level4">
<h4 class="anchored" data-anchor-id="positive-regions-are-defined-this-way">Positive regions are defined this way</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>motif_instances <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="fu">glue</span>(<span class="st">'{output_dir}/{id}_{TF}_{cell_line}_findinstances.txt'</span>), <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> T, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>motif_instances</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   PositionID Offset   Sequence  Motif.Name Strand MotifScore
1     peak153     45 CATGTAAATA FOXA1_LuCaP      +   6.653926
2      peak52     91 TGAATAAACA FOXA1_LuCaP      +   8.522032
3      peak26     67 TGTTTGCTGT FOXA1_LuCaP      -   7.450060
4     peak133     54 TGTTTACTTT FOXA1_LuCaP      -  11.429705
5      peak27    169 TATTTACATA FOXA1_LuCaP      -   9.571697
6     peak180     68 TAAGTAAATA FOXA1_LuCaP      +  10.179864
7      peak28     53 ACAGCAAACA FOXA1_LuCaP      +   7.450060
8      peak12     47 AAAACAAACA FOXA1_LuCaP      +   7.796603
9      peak41     41 AGAGCAAATA FOXA1_LuCaP      +   7.480315
10    peak108    157 TGTTGACTTT FOXA1_LuCaP      -   9.289078
11    peak181     68 TATTGACATA FOXA1_LuCaP      -   7.431071
12    peak193     53 CAAATAAACA FOXA1_LuCaP      +   6.591322
13    peak103     10 TGTTTGCTTA FOXA1_LuCaP      -   9.708916
14    peak163     18 AAAGTAAACA FOXA1_LuCaP      +  11.429705
15    peak130     60 TGTTTGCATT FOXA1_LuCaP      -   9.104899
16     peak86     15 ATTGTAAATA FOXA1_LuCaP      +   6.999968
17    peak111     35 TGTTTACCTA FOXA1_LuCaP      -   9.780715</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>chip_regions <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="fu">glue</span>(<span class="st">'{output_dir}/{id}_{TF}_{cell_line}_train_df.txt'</span>), <span class="at">sep=</span><span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">header=</span>F, <span class="at">stringsAsFactors=</span>F) </span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>motif_library <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="fu">glue</span>(<span class="st">'{common_dir}/{TF}_{cell_line}_motif_threshold_info.txt'</span>), <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> F, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(chip_regions) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"seqnames"</span>,<span class="st">"starts"</span>,<span class="st">"ends"</span>,<span class="st">"names"</span>,<span class="st">"scores"</span>,<span class="st">"strands"</span>)</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="co">#select ChIP peaks with a matching motif under them.</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>chip_regions_full <span class="ot">&lt;-</span> chip_regions</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">match</span>(chip_regions[,<span class="dv">4</span>], motif_instances[,<span class="dv">1</span>]) <span class="co"># what chips are in motifs data?</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>chip_regions <span class="ot">&lt;-</span> chip_regions_full[<span class="fu">which</span>(<span class="fu">is.na</span>(m) <span class="sc">==</span> F),]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>chip_regions_newpeaks <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>,<span class="fu">nrow</span>(chip_regions),<span class="dv">6</span>) <span class="co">#dummy var number of rows </span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(chip_regions)){</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"row "</span>, i, <span class="st">" of "</span>, <span class="fu">nrow</span>(chip_regions)))</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>    w <span class="ot">&lt;-</span> <span class="fu">which</span>(motif_instances<span class="sc">$</span>PositionID <span class="sc">==</span> chip_regions[i,<span class="dv">4</span>]) <span class="co">#match peakID, gauranteed to have a match </span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>    <span class="co">#only choose 1 motif, take the one with the top score</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>    w1 <span class="ot">&lt;-</span> w[<span class="fu">match</span>(<span class="fu">max</span>(motif_instances<span class="sc">$</span>MotifScore[w]), motif_instances<span class="sc">$</span>MotifScore[w])]</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (motif_instances<span class="sc">$</span>Strand[w1] <span class="sc">==</span> <span class="st">"+"</span>){</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>        centerpeak <span class="ot">&lt;-</span> motif_instances<span class="sc">$</span>Offset[w1] <span class="sc">+</span> <span class="fu">floor</span>(<span class="fl">0.5</span><span class="sc">*</span>(<span class="fu">nchar</span>(motif_instances<span class="sc">$</span>Sequence[w1]))) <span class="co">#this is relative to start </span></span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>    }<span class="cf">else</span>{</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>        centerpeak <span class="ot">&lt;-</span> motif_instances<span class="sc">$</span>Offset[w1] <span class="sc">-</span> <span class="fu">floor</span>(<span class="fl">0.5</span><span class="sc">*</span>(<span class="fu">nchar</span>(motif_instances<span class="sc">$</span>Sequence[w1])))</span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a>    newstart <span class="ot">&lt;-</span> chip_regions<span class="sc">$</span>starts[i]<span class="sc">+</span>centerpeak <span class="co">#previous version of impact -windowsize/2 #establish left bound of peaks</span></span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>    newend <span class="ot">&lt;-</span> newstart <span class="sc">+</span> <span class="dv">1</span> <span class="co">#previous version of impact +windowsize #establish right bound of peaks </span></span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a>    chip_regions_newpeaks[i,<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="fu">as.character</span>(chip_regions<span class="sc">$</span>seqnames[i])</span>
<span id="cb33-18"><a href="#cb33-18" aria-hidden="true" tabindex="-1"></a>    chip_regions_newpeaks[i,<span class="dv">2</span>] <span class="ot">&lt;-</span> newstart</span>
<span id="cb33-19"><a href="#cb33-19" aria-hidden="true" tabindex="-1"></a>    chip_regions_newpeaks[i,<span class="dv">3</span>] <span class="ot">&lt;-</span> newend</span>
<span id="cb33-20"><a href="#cb33-20" aria-hidden="true" tabindex="-1"></a>    chip_regions_newpeaks[i,<span class="dv">4</span>] <span class="ot">&lt;-</span> <span class="fu">as.character</span>(chip_regions<span class="sc">$</span>names[i]) <span class="co">#preserve old information</span></span>
<span id="cb33-21"><a href="#cb33-21" aria-hidden="true" tabindex="-1"></a>    chip_regions_newpeaks[i,<span class="dv">5</span>] <span class="ot">&lt;-</span> <span class="fu">as.character</span>(chip_regions<span class="sc">$</span>scores[i])</span>
<span id="cb33-22"><a href="#cb33-22" aria-hidden="true" tabindex="-1"></a>    chip_regions_newpeaks[i,<span class="dv">6</span>] <span class="ot">&lt;-</span> <span class="fu">as.character</span>(chip_regions<span class="sc">$</span>strands[i])</span>
<span id="cb33-23"><a href="#cb33-23" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "row 1 of 17"
[1] "row 2 of 17"
[1] "row 3 of 17"
[1] "row 4 of 17"
[1] "row 5 of 17"
[1] "row 6 of 17"
[1] "row 7 of 17"
[1] "row 8 of 17"
[1] "row 9 of 17"
[1] "row 10 of 17"
[1] "row 11 of 17"
[1] "row 12 of 17"
[1] "row 13 of 17"
[1] "row 14 of 17"
[1] "row 15 of 17"
[1] "row 16 of 17"
[1] "row 17 of 17"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(chip_regions_newpeaks, <span class="fu">glue</span>(<span class="st">"{output_dir}/{id}_{TF}_{cell_line}_train_df_newpeaks.txt"</span>), <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">quote =</span> F, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>chip_regions_newpeaks</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      [,1]    [,2]        [,3]        [,4]      [,5] [,6]
 [1,] "chr6"  "71571271"  "71571272"  "peak12"  "0"  "+" 
 [2,] "chr18" "77901662"  "77901663"  "peak26"  "0"  "+" 
 [3,] "chr15" "77278094"  "77278095"  "peak27"  "0"  "+" 
 [4,] "chr8"  "42748581"  "42748582"  "peak28"  "0"  "+" 
 [5,] "chr6"  "57069709"  "57069710"  "peak41"  "0"  "+" 
 [6,] "chr20" "52894395"  "52894396"  "peak52"  "0"  "+" 
 [7,] "chr1"  "180609383" "180609384" "peak86"  "0"  "+" 
 [8,] "chr2"  "163281248" "163281249" "peak103" "0"  "+" 
 [9,] "chr6"  "19754304"  "19754305"  "peak108" "0"  "+" 
[10,] "chr1"  "166217851" "166217852" "peak111" "0"  "+" 
[11,] "chr2"  "1468954"   "1468955"   "peak130" "0"  "+" 
[12,] "chr16" "3271116"   "3271117"   "peak133" "0"  "+" 
[13,] "chrX"  "152781622" "152781623" "peak153" "0"  "+" 
[14,] "chr2"  "152630569" "152630570" "peak163" "0"  "+" 
[15,] "chr12" "42040509"  "42040510"  "peak180" "0"  "+" 
[16,] "chr3"  "67088159"  "67088160"  "peak181" "0"  "+" 
[17,] "chr3"  "66479831"  "66479832"  "peak193" "0"  "+" </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>chip_regions_newpeaks <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="fu">glue</span>(<span class="st">'{output_dir}/{id}_{TF}_{cell_line}_train_df_newpeaks.txt'</span>), <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> F, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>chip_regions_newpeaks_realdf <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">seqnames=</span>chip_regions_newpeaks[,<span class="dv">1</span>],</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">starts=</span><span class="fu">as.numeric</span>(chip_regions_newpeaks[,<span class="dv">2</span>]),</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">ends=</span><span class="fu">as.numeric</span>(chip_regions_newpeaks[,<span class="dv">3</span>]),</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">names=</span>chip_regions_newpeaks[,<span class="dv">4</span>],</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">scores=</span><span class="dv">0</span>,</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">strands=</span><span class="st">"+"</span>)</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2022</span>)</span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>s <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">seq</span>(<span class="dv">1</span>,<span class="fu">nrow</span>(chip_regions_newpeaks_realdf),<span class="dv">1</span>), <span class="fu">nrow</span>(chip_regions_newpeaks_realdf),<span class="at">replace =</span> F)</span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>s</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  4  3 14  7 11 17  6  9  8 15  5  2 10  1 16 13 12</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>chip_regions_newpeaks_realdf <span class="ot">&lt;-</span> chip_regions_newpeaks_realdf[s,]</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="co">#put training and test together! </span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="co">#use partition function later to make it easier to sample many times! </span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>train_test_peaks <span class="ot">&lt;-</span> <span class="fu">unique</span>(chip_regions_newpeaks_realdf)</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(train_test_peaks) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"seqnames"</span>,<span class="st">"starts"</span>,<span class="st">"ends"</span>,<span class="st">"names"</span>,<span class="st">"scores"</span>,<span class="st">"strands"</span>)</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a><span class="co">#clean up </span></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(<span class="fu">which</span>(train_test_peaks<span class="sc">$</span>start <span class="sc">&lt;</span> <span class="dv">0</span>)) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>    train_test_peaks<span class="sc">$</span>start[<span class="fu">which</span>(train_test_peaks<span class="sc">$</span>start <span class="sc">&lt;</span> <span class="dv">0</span>)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>t <span class="ot">&lt;-</span> train_test_peaks[,<span class="dv">3</span>]<span class="sc">-</span>train_test_peaks[,<span class="dv">2</span>]</span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>w <span class="ot">&lt;-</span> <span class="fu">which</span>(t <span class="sc">%%</span> <span class="dv">2</span> <span class="sc">!=</span> <span class="dv">0</span>) <span class="co">#odd</span></span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(w) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a>    train_test_peaks[w,<span class="dv">3</span>] <span class="ot">&lt;-</span> train_test_peaks[w,<span class="dv">3</span>] <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb39-17"><a href="#cb39-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-18"><a href="#cb39-18" aria-hidden="true" tabindex="-1"></a>w1 <span class="ot">&lt;-</span> <span class="fu">which</span>(train_test_peaks<span class="sc">$</span>chr <span class="sc">==</span> <span class="st">"chrM"</span>)</span>
<span id="cb39-19"><a href="#cb39-19" aria-hidden="true" tabindex="-1"></a>w2 <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">nchar</span>(train_test_peaks<span class="sc">$</span>chr) <span class="sc">&gt;</span> <span class="dv">5</span>)</span>
<span id="cb39-20"><a href="#cb39-20" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(<span class="fu">c</span>(w1,w2)) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb39-21"><a href="#cb39-21" aria-hidden="true" tabindex="-1"></a>    train_test_peaks <span class="ot">&lt;-</span> train_test_peaks[<span class="sc">-</span><span class="fu">c</span>(w1,w2),]</span>
<span id="cb39-22"><a href="#cb39-22" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb39-23"><a href="#cb39-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-24"><a href="#cb39-24" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(train_test_peaks, <span class="fu">glue</span>(<span class="st">"{output_dir}/{id}_{TF}_{cell_line}_train_test_positive_bed.txt"</span>), <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">quote =</span> F, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">col.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb39-25"><a href="#cb39-25" aria-hidden="true" tabindex="-1"></a>train_test_peaks</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   seqnames    starts      ends   names scores strands
4      chr8  42748581  42748583  peak28      0       +
3     chr15  77278094  77278096  peak27      0       +
14     chr2 152630569 152630571 peak163      0       +
7      chr1 180609383 180609385  peak86      0       +
11     chr2   1468954   1468956 peak130      0       +
17     chr3  66479831  66479833 peak193      0       +
6     chr20  52894395  52894397  peak52      0       +
9      chr6  19754304  19754306 peak108      0       +
8      chr2 163281248 163281250 peak103      0       +
15    chr12  42040509  42040511 peak180      0       +
5      chr6  57069709  57069711  peak41      0       +
2     chr18  77901662  77901664  peak26      0       +
10     chr1 166217851 166217853 peak111      0       +
1      chr6  71571271  71571273  peak12      0       +
16     chr3  67088159  67088161 peak181      0       +
13     chrX 152781622 152781624 peak153      0       +
12    chr16   3271116   3271118 peak133      0       +</code></pre>
</div>
</div>
</section>
<section id="negative-sets-are-defined-this-way" class="level4">
<h4 class="anchored" data-anchor-id="negative-sets-are-defined-this-way">Negative sets are defined this way</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>train_df_newpeaks <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="fu">glue</span>(<span class="st">"{output_dir}/{id}_{TF}_{cell_line}_train_df_newpeaks.txt"</span>), <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> F, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>MotifThreshold <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="fu">glue</span>(<span class="st">"{common_dir}/{TF}_{cell_line}_motif_threshold_info.txt"</span>), <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> F, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>motifs <span class="ot">&lt;-</span> MotifThreshold[<span class="dv">1</span>,<span class="dv">1</span>]</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>instscan <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="fu">glue</span>(<span class="st">"{common_dir}/{TF}_{cell_line}_scanMotifsGenomewide_sort_15000.txt"</span>), <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> F, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>instscan[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   V1    V2       V3       V4 V5       V6         V7
1 FOXA1_LuCaP-1000025 chr19 44523565 44523574  + 11.42971 AAAGTAAACA
2 FOXA1_LuCaP-1000078 chr19 44580731 44580740  - 11.42971 TGTTTACTTT
3 FOXA1_LuCaP-1000084 chr19 44594356 44594365  + 11.42971 AAAGTAAACA
4 FOXA1_LuCaP-1000278 chr19 44772423 44772432  - 11.42971 TGTTTACTTT
5 FOXA1_LuCaP-1000957 chr19 45700582 45700591  + 11.42971 AAAGTAAACA</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>instscan <span class="ot">&lt;-</span> instscan[<span class="fu">order</span>(instscan[,<span class="dv">6</span>], <span class="at">decreasing =</span> T), ]   </span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>train_df <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="fu">glue</span>(<span class="st">"{output_dir}/{id}_{TF}_{cell_line}_train_df.txt"</span>), <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> F, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>) </span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(train_df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'chr'</span>,<span class="st">'start'</span>,<span class="st">'end'</span>,<span class="st">'id'</span>,<span class="st">'score'</span>,<span class="st">'strand'</span>) <span class="co">#score may be number of reads</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>Granges <span class="ot">&lt;-</span> <span class="fu">with</span>(train_df, <span class="fu">GRanges</span>(chr, <span class="fu">IRanges</span>(start,end), strand, score, <span class="at">id =</span> id))</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="fu">assign</span>(<span class="fu">paste0</span>(motifs,<span class="st">"_Granges"</span>),Granges)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>allinstscan <span class="ot">&lt;-</span> instscan</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>sapply_instscan <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(allinstscan), <span class="cf">function</span>(x) <span class="fu">strsplit</span>(allinstscan[x,<span class="dv">1</span>], <span class="st">"-"</span>)[[<span class="dv">1</span>]][<span class="dv">1</span>])</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>allinstscan[,<span class="dv">1</span>] <span class="ot">&lt;-</span> sapply_instscan</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>allinstscan[,<span class="dv">6</span>] <span class="ot">&lt;-</span> allinstscan[,<span class="dv">6</span>]<span class="sc">/</span>MotifThreshold[<span class="dv">1</span>,<span class="dv">2</span>]</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>allinstscan[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           V1    V2       V3       V4 V5       V6         V7
1 FOXA1_LuCaP chr19 44523565 44523574  + 1.803496 AAAGTAAACA
2 FOXA1_LuCaP chr19 44580731 44580740  - 1.803496 TGTTTACTTT
3 FOXA1_LuCaP chr19 44594356 44594365  + 1.803496 AAAGTAAACA
4 FOXA1_LuCaP chr19 44772423 44772432  - 1.803496 TGTTTACTTT
5 FOXA1_LuCaP chr19 45700582 45700591  + 1.803496 AAAGTAAACA</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>w <span class="ot">&lt;-</span> <span class="fu">which</span>(allinstscan[,<span class="dv">6</span>] <span class="sc">&lt;</span> <span class="dv">1</span>) </span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(w)<span class="sc">&gt;</span><span class="dv">0</span>){allinstscan <span class="ot">&lt;-</span> allinstscan[<span class="sc">-</span>w,]}</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>motifmatrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>,<span class="fu">nrow</span>(MotifThreshold),<span class="dv">3</span>)</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>motifmatrix[,<span class="dv">1</span>] <span class="ot">&lt;-</span> motifs </span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>motifmatrix[,<span class="dv">2</span>] <span class="ot">&lt;-</span> motifs</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>motifmatrix[,<span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="fu">paste0</span>(motifs,<span class="st">"_Granges"</span>)</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>motifmatrix</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1]          [,2]          [,3]                 
[1,] "FOXA1_LuCaP" "FOXA1_LuCaP" "FOXA1_LuCaP_Granges"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"working on "</span>,motifmatrix[<span class="dv">1</span>,<span class="dv">1</span>]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "working on FOXA1_LuCaP"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>w_mcp <span class="ot">&lt;-</span> <span class="fu">which</span>(train_df_newpeaks[,<span class="dv">5</span>] <span class="sc">==</span> motifmatrix[<span class="dv">1</span>,<span class="dv">2</span>]) <span class="co">#motifcentered peaks</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>motifcentered_specificTF <span class="ot">&lt;-</span> train_df_newpeaks[w_mcp,]</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(motifcentered_specificTF) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'chr'</span>,<span class="st">'start'</span>,<span class="st">'end'</span>,<span class="st">'id'</span>,<span class="st">'score'</span>,<span class="st">'strand'</span>) </span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>Granges_obj_mc <span class="ot">&lt;-</span> <span class="fu">with</span>(motifcentered_specificTF, <span class="fu">GRanges</span>(chr, <span class="fu">IRanges</span>(start,end), strand, score, <span class="at">id =</span> id))</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>Granges_obj_mc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 0 ranges and 2 metadata columns:
   seqnames    ranges strand |     score          id
      &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
  -------
  seqinfo: no sequences</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>w_inst <span class="ot">&lt;-</span> <span class="fu">which</span>(allinstscan[,<span class="dv">1</span>] <span class="sc">==</span> motifmatrix[<span class="dv">1</span>,<span class="dv">2</span>]) <span class="co">#instances pertaining to specific motif</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>instances_specificTF <span class="ot">&lt;-</span> allinstscan[w_inst,]</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>inst_bed <span class="ot">&lt;-</span> instances_specificTF[,<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>)]</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>inst_bed <span class="ot">&lt;-</span> <span class="fu">cbind</span>(inst_bed, motifmatrix[<span class="dv">1</span>,<span class="dv">1</span>], instances_specificTF[,<span class="dv">6</span>], <span class="st">"+"</span>) </span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a><span class="co">#region as wide as a motif </span></span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>motif_width <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(inst_bed[,<span class="dv">3</span>]) <span class="sc">-</span> <span class="fu">as.numeric</span>(inst_bed[,<span class="dv">2</span>])</span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>inst_bed[,<span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(inst_bed[,<span class="dv">2</span>]) <span class="sc">+</span> <span class="fu">floor</span>(motif_width<span class="sc">/</span><span class="dv">2</span>)</span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a>inst_bed[,<span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(inst_bed[,<span class="dv">2</span>]) <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a><span class="co">#inst_bed[,3] &lt;- as.numeric(inst_bed[,3])#previous version of impact: +(windowsize-10)/2</span></span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a><span class="co">#inst_bed[,2] &lt;- as.numeric(inst_bed[,2])#previous version of impact: -(windowsize-10)/2</span></span>
<span id="cb53-13"><a href="#cb53-13" aria-hidden="true" tabindex="-1"></a>inst_bed <span class="ot">&lt;-</span> <span class="fu">unique</span>(inst_bed)</span>
<span id="cb53-14"><a href="#cb53-14" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(inst_bed) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'chr'</span>,<span class="st">'start'</span>,<span class="st">'end'</span>,<span class="st">'id'</span>,<span class="st">'score'</span>,<span class="st">'strand'</span>) </span>
<span id="cb53-15"><a href="#cb53-15" aria-hidden="true" tabindex="-1"></a>inst_bed_Granges <span class="ot">&lt;-</span> <span class="fu">with</span>(inst_bed, <span class="fu">GRanges</span>(chr, <span class="fu">IRanges</span>(start,end), strand, score, <span class="at">id =</span> id)) </span>
<span id="cb53-16"><a href="#cb53-16" aria-hidden="true" tabindex="-1"></a>inst_bed_Granges</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 15000 ranges and 2 metadata columns:
          seqnames              ranges strand |     score          id
             &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;character&gt;
      [1]    chr19   44523569-44523570      + |    1.8035 FOXA1_LuCaP
      [2]    chr19   44580735-44580736      + |    1.8035 FOXA1_LuCaP
      [3]    chr19   44594360-44594361      + |    1.8035 FOXA1_LuCaP
      [4]    chr19   44772427-44772428      + |    1.8035 FOXA1_LuCaP
      [5]    chr19   45700586-45700587      + |    1.8035 FOXA1_LuCaP
      ...      ...                 ...    ... .       ...         ...
  [14996]     chr7 126821777-126821778      + |    1.8035 FOXA1_LuCaP
  [14997]     chr7 126853810-126853811      + |    1.8035 FOXA1_LuCaP
  [14998]     chr7 126935076-126935077      + |    1.8035 FOXA1_LuCaP
  [14999]     chr7 127088380-127088381      + |    1.8035 FOXA1_LuCaP
  [15000]     chr7 127103853-127103854      + |    1.8035 FOXA1_LuCaP
  -------
  seqinfo: 25 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>int_1 <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">intersect</span>(inst_bed_Granges, <span class="fu">get</span>(motifmatrix[<span class="dv">1</span>,<span class="dv">3</span>]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in .Seqinfo.mergexy(x, y): Each of the 2 combined objects has sequence levels not in the other:
  - in 'x': chr1_gl000191_random, chr1_gl000192_random, chr4_ctg9_hap1, chr4_gl000193_random, chr4_gl000194_random, chr6_apd_hap1, chr6_cox_hap2, chr6_dbb_hap3, chr6_mann_hap4, chr6_mcf_hap5, chr6_qbl_hap6, chr6_ssto_hap7
  - in 'y': chrX, chr14, chr12, chr16, chr9, chr18, chr15, chr8, chr13, chr17
  Make sure to always combine/compare objects based on the same reference
  genome (use suppressWarnings() to suppress this warning).</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>int_2 <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">intersect</span>(inst_bed_Granges, Granges_obj_mc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Also the same as below (…)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>overlaps <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">findOverlaps</span>(inst_bed_Granges, <span class="fu">get</span>(motifmatrix[<span class="dv">1</span>,<span class="dv">3</span>]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in .Seqinfo.mergexy(x, y): Each of the 2 combined objects has sequence levels not in the other:
  - in 'x': chr1_gl000191_random, chr1_gl000192_random, chr4_ctg9_hap1, chr4_gl000193_random, chr4_gl000194_random, chr6_apd_hap1, chr6_cox_hap2, chr6_dbb_hap3, chr6_mann_hap4, chr6_mcf_hap5, chr6_qbl_hap6, chr6_ssto_hap7
  - in 'y': chrX, chr14, chr12, chr16, chr9, chr18, chr15, chr8, chr13, chr17
  Make sure to always combine/compare objects based on the same reference
  genome (use suppressWarnings() to suppress this warning).</code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>overlaps ; inst_bed_Granges[<span class="fu">queryHits</span>(overlaps), ] ; <span class="fu">get</span>(motifmatrix[<span class="dv">1</span>,<span class="dv">3</span>])[<span class="fu">subjectHits</span>(overlaps), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Hits object with 1 hit and 0 metadata columns:
      queryHits subjectHits
      &lt;integer&gt;   &lt;integer&gt;
  [1]      4724         163
  -------
  queryLength: 15000 / subjectLength: 200</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 1 range and 2 metadata columns:
      seqnames              ranges strand |     score          id
         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;character&gt;
  [1]     chr2 152630569-152630570      + |    1.8035 FOXA1_LuCaP
  -------
  seqinfo: 25 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 1 range and 2 metadata columns:
      seqnames              ranges strand |     score          id
         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
  [1]     chr2 152630546-152630653      + |         0     peak163
  -------
  seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>Continue…</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>ms_1 <span class="ot">&lt;-</span> <span class="fu">match</span>(<span class="fu">start</span>(int_1)<span class="sc">-</span><span class="dv">1</span>, inst_bed<span class="sc">$</span>start)</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>me_1 <span class="ot">&lt;-</span> <span class="fu">match</span>(<span class="fu">end</span>(int_1), inst_bed<span class="sc">$</span>end)</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">c</span>(ms_1,me_1)</span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">unique</span>(a[<span class="fu">is.na</span>(a) <span class="sc">==</span> F])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>ms_2 <span class="ot">&lt;-</span> <span class="fu">match</span>(<span class="fu">start</span>(int_2)<span class="sc">-</span><span class="dv">1</span>, inst_bed<span class="sc">$</span>start)</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>me_2 <span class="ot">&lt;-</span> <span class="fu">match</span>(<span class="fu">end</span>(int_2), inst_bed<span class="sc">$</span>end)</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="fu">c</span>(ms_2,me_2)</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="fu">unique</span>(b[<span class="fu">is.na</span>(b) <span class="sc">==</span> F])</span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>ab <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(a,b))</span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>ab <span class="ot">&lt;-</span> ab[<span class="sc">!</span><span class="fu">is.na</span>(ab)]</span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>ab</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4724</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">length</span>(ab) <span class="sc">&gt;</span> <span class="dv">0</span>){</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>    noChIP <span class="ot">&lt;-</span> inst_bed[<span class="sc">-</span>ab,] <span class="co">#good negative peaks</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>}<span class="cf">else</span>{</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>    noChIP <span class="ot">&lt;-</span> inst_bed</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(noChIP)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 14999     6</code></pre>
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
<section id="my-approach" class="level2">
<h2 class="anchored" data-anchor-id="my-approach">My approach</h2>
<p>Remember I have the <code>chip_granges</code> and the top 15000 predicted motifs</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>chip_granges</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 200 ranges and 0 metadata columns:
        seqnames              ranges strand
           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;
    [1]     chrX 103122567-103122585      *
    [2]     chr6   37842908-37842911      *
    [3]     chr5 142305491-142305568      *
    [4]     chr1 164453204-164453206      *
    [5]     chr6   42922199-42922204      *
    ...      ...                 ...    ...
  [196]    chr18   74902999-74903006      *
  [197]     chr2 144059425-144059465      *
  [198]    chr13 114548061-114548066      *
  [199]     chr6   44526815-44526818      *
  [200]     chr6 144736546-144736560      *
  -------
  seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>pred_motifs_granges <span class="ot">&lt;-</span> <span class="fu">with</span>(instscan, <span class="fu">GRanges</span>(V2, <span class="fu">IRanges</span>(V3,V4), <span class="at">strand=</span>V5, <span class="at">score=</span>V6))</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>pred_motifs_granges</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 15000 ranges and 1 metadata column:
          seqnames              ranges strand |     score
             &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
      [1]    chr19   44523565-44523574      + |   11.4297
      [2]    chr19   44580731-44580740      - |   11.4297
      [3]    chr19   44594356-44594365      + |   11.4297
      [4]    chr19   44772423-44772432      - |   11.4297
      [5]    chr19   45700582-45700591      + |   11.4297
      ...      ...                 ...    ... .       ...
  [14996]     chr7 126821773-126821782      + |   11.4297
  [14997]     chr7 126853806-126853815      - |   11.4297
  [14998]     chr7 126935072-126935081      + |   11.4297
  [14999]     chr7 127088376-127088385      + |   11.4297
  [15000]     chr7 127103849-127103858      - |   11.4297
  -------
  seqinfo: 25 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>My approach is to simply overlap them</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>overlaps <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">findOverlaps</span>(<span class="at">query=</span>chip_granges, <span class="at">subject=</span>pred_motifs_granges,<span class="at">type=</span><span class="st">'any'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in .Seqinfo.mergexy(x, y): Each of the 2 combined objects has sequence levels not in the other:
  - in 'x': chrX, chr14, chr12, chr16, chr9, chr18, chr15, chr8, chr13, chr17
  - in 'y': chr1_gl000191_random, chr1_gl000192_random, chr4_ctg9_hap1, chr4_gl000193_random, chr4_gl000194_random, chr6_apd_hap1, chr6_cox_hap2, chr6_dbb_hap3, chr6_mann_hap4, chr6_mcf_hap5, chr6_qbl_hap6, chr6_ssto_hap7
  Make sure to always combine/compare objects based on the same reference
  genome (use suppressWarnings() to suppress this warning).</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>overlaps</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Hits object with 1 hit and 0 metadata columns:
      queryHits subjectHits
      &lt;integer&gt;   &lt;integer&gt;
  [1]       163        4724
  -------
  queryLength: 200 / subjectLength: 15000</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>chip_granges[<span class="fu">queryHits</span>(overlaps), ] ; pred_motifs_granges[<span class="fu">subjectHits</span>(overlaps), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 1 range and 0 metadata columns:
      seqnames              ranges strand
         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt;
  [1]     chr2 152630546-152630653      *
  -------
  seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 1 range and 1 metadata column:
      seqnames              ranges strand |     score
         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
  [1]     chr2 152630565-152630574      + |   11.4297
  -------
  seqinfo: 25 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>I am getting, qualitatively, the same result.</p>
<p>But if I do it my way, I will choose just this region as a bound region.</p>
<p>And these remaining regions as my unbound regions</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>pred_motifs_granges[<span class="sc">-</span><span class="fu">subjectHits</span>(overlaps), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 14999 ranges and 1 metadata column:
          seqnames              ranges strand |     score
             &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
      [1]    chr19   44523565-44523574      + |   11.4297
      [2]    chr19   44580731-44580740      - |   11.4297
      [3]    chr19   44594356-44594365      + |   11.4297
      [4]    chr19   44772423-44772432      - |   11.4297
      [5]    chr19   45700582-45700591      + |   11.4297
      ...      ...                 ...    ... .       ...
  [14995]     chr7 126821773-126821782      + |   11.4297
  [14996]     chr7 126853806-126853815      - |   11.4297
  [14997]     chr7 126935072-126935081      + |   11.4297
  [14998]     chr7 127088376-127088385      + |   11.4297
  [14999]     chr7 127103849-127103858      - |   11.4297
  -------
  seqinfo: 25 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>