#!/bin/bash
#PBS -l select=1:system=polaris
#PBS -l walltime=01:00:00,filesystems=home:grand
#PBS -A covid-ct
#PBS -q debug   
#PBS -N select_snps
#PBS -k doe
#PBS -o /lus/grand/projects/TFXcan/imlab/users/temi/projects/TFXcan/logs/select_snps.out
#PBS -e /lus/grand/projects/TFXcan/imlab/users/temi/projects/TFXcan/logs/select_snps.err

echo Working directory is $PBS_O_WORKDIR
cd $PBS_O_WORKDIR

source ~/.bashrc
conda activate /lus/grand/projects/TFXcan/imlab/users/temi/software/conda_envs/compbio-tools
module load gnu-parallel

fasta_file=/lus/grand/projects/TFXcan/imlab/data/hg_sequences/hg38/Homo_sapiens_assembly38.fasta
input_dir=/lus/grand/projects/TFXcan/imlab/data/baca_cwas/imputed_vcf_hg38
output_dir=/lus/grand/projects/TFXcan/imlab/data/baca_cwas/imputed_vcf_hg38_snps_only
mkdir -p ${output_dir}

# parallel -j 22 "bcftools view -r chr{} --output-type z --output ${output_dir}/chr{}_CWAS_SNPs_2023-05-30_phased.vcf.gz ${input_dir}/chr{}_CWAS_SNPs_2023-05-15_phased.vcf.gz && tabix -p vcf ${output_dir}/chr{}_CWAS_SNPs_2023-05-30_phased.vcf.gz"

seq 1 22 | parallel -j 4 "gatk SelectVariants -R ${fasta_file} -V ${input_dir}/chr{}.dose.vcf.gz --select-type-to-include SNP -O ${output_dir}/chr{}.dose.vcf.gz && tabix -p vcf ${output_dir}/chr{}.dose.vcf.gz"