<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.554">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Temi">
<meta name="dcterms.date" content="2024-09-11">
<meta name="description" content="This notebook contains steps and codes to reproduce, as much as possible, the Enpact paper results">

<title>Reproducing the Enpact data and results</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="recreateEnpactResults_files/libs/clipboard/clipboard.min.js"></script>
<script src="recreateEnpactResults_files/libs/quarto-html/quarto.js"></script>
<script src="recreateEnpactResults_files/libs/quarto-html/popper.min.js"></script>
<script src="recreateEnpactResults_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="recreateEnpactResults_files/libs/quarto-html/anchor.min.js"></script>
<link href="recreateEnpactResults_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="recreateEnpactResults_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="recreateEnpactResults_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="recreateEnpactResults_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="recreateEnpactResults_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Reproducing the Enpact data and results</h1>
</div>

<div>
  <div class="description">
    This notebook contains steps and codes to reproduce, as much as possible, the Enpact paper results
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Temi </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">September 11, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span>opts_chunk<span class="sc">$</span><span class="fu">set</span>(<span class="at">eval =</span> F, <span class="at">echo =</span> <span class="cn">TRUE</span>, <span class="at">message =</span> <span class="cn">TRUE</span>, <span class="at">warning =</span> <span class="cn">TRUE</span>, <span class="at">cache =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">'/project/haky/users/temi/projects/Enpact/notebooks'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glue)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RSQLite)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rtracklayer)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(yaml)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(fpeek)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(purrr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tdate &lt;- '2023-08-17' # modified</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>tdate <span class="ot">&lt;-</span> <span class="st">'2024-04-17'</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>base_dir <span class="ot">&lt;-</span> <span class="st">'/project2/haky/temi/projects/TFXcan/baca_cwas'</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>data_dir <span class="ot">&lt;-</span> <span class="st">'/project/haky/users/temi/projects/Enpact'</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>output_dir <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'{base_dir}/data'</span>)</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>files_dir <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'{data_dir}/files'</span>)</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span>(<span class="fu">dir.exists</span>(files_dir))){</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dir.create</span>(files_dir, <span class="at">recursive =</span> T)</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="introduction" class="level1">
<h1>Introduction</h1>
<p>In this notebook, I detail steps to re-create some of the data/results in the Enpact paper. I have set this notebook not to run when rendering. So, it will show you the code and markdown text, but not run the code.</p>
</section>
<section id="train-dl-based-enpact-models-from-scratch" class="level1">
<h1>Train DL-based Enpact models from scratch</h1>
<section id="prepare-the-metadata-and-files" class="level2">
<h2 class="anchored" data-anchor-id="prepare-the-metadata-and-files">Prepare the metadata and files</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>homerdb <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="fu">file.path</span>(data_dir, <span class="st">'metadata'</span>, <span class="st">'motifTable.txt'</span>), <span class="at">header =</span> T) <span class="sc">%&gt;%</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">','</span>, <span class="st">`</span><span class="at">Gene Symbol</span><span class="st">`</span>), <span class="sc">!</span><span class="st">`</span><span class="at">Gene Symbol</span><span class="st">`</span> <span class="sc">%in%</span> <span class="fu">c</span>( <span class="st">"-"</span>, <span class="st">"?"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(Filename, Consensus, <span class="at">symbol=</span><span class="st">`</span><span class="at">Gene Symbol</span><span class="st">`</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(symbol <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">'AR'</span>, <span class="st">'FOXA1'</span>, <span class="st">'GATA2'</span>, <span class="st">'HOXB13'</span>))</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>homerdb</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>mt <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="fu">file.path</span>(data_dir, <span class="st">'metadata'</span>, <span class="st">'human_factor_full_QC.txt'</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(Tissue_type <span class="sc">!=</span> <span class="st">'None'</span>, <span class="sc">!</span><span class="fu">is.na</span>(PeaksUnionDHSRatio), FRiP <span class="sc">&gt;</span> <span class="fl">0.01</span>) <span class="sc">%&gt;%</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">'-'</span>, Factor, <span class="at">fixed =</span> T)) <span class="sc">%&gt;%</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(Factor <span class="sc">%in%</span> homerdb<span class="sc">$</span>symbol)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>idt <span class="ot">&lt;-</span> mt <span class="sc">%&gt;%</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(Factor <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">'AR'</span>, <span class="st">'FOXA1'</span>, <span class="st">'GATA2'</span>, <span class="st">'HOXB13'</span>), Tissue_type <span class="sc">==</span> <span class="st">'Prostate'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(Factor, Tissue_type) <span class="sc">%&gt;%</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_split</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>ldt <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">seq_along</span>(idt)){</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    fdt <span class="ot">&lt;-</span> idt[[i]]</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    tissuename <span class="ot">&lt;-</span> <span class="fu">unique</span>(fdt<span class="sc">$</span>Tissue_type)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>    tissuename <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">' '</span>, <span class="st">''</span>, tissuename)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    tfname <span class="ot">&lt;-</span> <span class="fu">unique</span>(fdt<span class="sc">$</span>Factor)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    dcids <span class="ot">&lt;-</span> fdt<span class="sc">$</span>DCid</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="sc">!</span>tfname <span class="sc">%in%</span> <span class="fu">names</span>(ldt)){</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>        ldt[[<span class="fu">as.name</span>(tfname)]] <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>        ldt[[<span class="fu">as.name</span>(tfname)]][[<span class="st">'peakFiles'</span>]] <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>        ldt[[<span class="fu">as.name</span>(tfname)]][[<span class="st">'peakFiles'</span>]][[<span class="fu">as.name</span>(tissuename)]] <span class="ot">&lt;-</span>  <span class="fu">paste0</span>(dcids, <span class="st">'_sort_peaks.narrowPeak.bed'</span>)</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>        ldt[[<span class="fu">as.name</span>(tfname)]][[<span class="st">'peakFiles'</span>]][[<span class="fu">as.name</span>(tissuename)]] <span class="ot">&lt;-</span>  <span class="fu">paste0</span>(dcids, <span class="st">'_sort_peaks.narrowPeak.bed'</span>)</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>factors_motifs <span class="ot">&lt;-</span> homerdb <span class="sc">%&gt;%</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(symbol) <span class="sc">%&gt;%</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_split</span>()</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>ndt <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>mdt <span class="ot">&lt;-</span> <span class="fu">lapply</span>(factors_motifs, <span class="cf">function</span>(edt){</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    res <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    res[[<span class="st">'motifFiles'</span>]] <span class="ot">&lt;-</span> edt<span class="sc">$</span>Filename</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>    ndt <span class="ot">&lt;&lt;-</span> <span class="fu">append</span>(ndt, <span class="fu">unique</span>(edt<span class="sc">$</span>symbol))</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(res)</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(mdt) <span class="ot">&lt;-</span> ndt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>common_names <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">names</span>(ldt), <span class="fu">names</span>(mdt))</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>peakslist <span class="ot">&lt;-</span> ldt[common_names]</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>motifslist <span class="ot">&lt;-</span> mdt[common_names]</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>enpact_models_config <span class="ot">&lt;-</span> <span class="fu">mapply</span>(c, peakslist, motifslist)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Modified from: https://stackoverflow.com/questions/74655073/how-to-effectively-join-two-lists-elementwise-by-element-name</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>cat_lists <span class="ot">&lt;-</span> <span class="cf">function</span>(list1, list2) {  </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  keys <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">names</span>(list1), <span class="fu">names</span>(list2)))</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">map2</span>(list1[keys], list2[keys], c) <span class="sc">|&gt;</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">set_names</span>(keys)  </span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>enlist <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">reduce</span>(<span class="fu">list</span>(motifslist, peakslist), cat_lists)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>– Filter for the following TFs: AR, FOXA1, HOXB13, and GATA2 in prostate tissues</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>tfs_ex <span class="ot">&lt;-</span> enlist[<span class="fu">c</span>(<span class="st">'AR'</span>, <span class="st">'FOXA1'</span>, <span class="st">'HOXB13'</span>, <span class="st">'GATA2'</span>)]</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>tfs_ex <span class="ot">&lt;-</span> base<span class="sc">::</span><span class="fu">Filter</span>(<span class="fu">Negate</span>(is.null), tfs_ex)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>yaml<span class="sc">::</span><span class="fu">write_yaml</span>(tfs_ex, <span class="at">file =</span> <span class="fu">file.path</span>(data_dir, <span class="st">'metadata'</span>, <span class="st">'enpact_models_to_train.yaml'</span>))</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co"># prepare the metadata </span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>mtdt <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">gsub</span>(<span class="st">'.peakFiles.'</span>, <span class="st">'_'</span>, <span class="fu">names</span>(<span class="fu">rapply</span>(peakslist, <span class="cf">function</span>(x) <span class="fu">head</span>(x, <span class="dv">1</span>)))), base<span class="sc">::</span>strsplit, <span class="st">'_'</span>) <span class="sc">|&gt;</span> <span class="fu">unname</span>() <span class="sc">%&gt;%</span> <span class="fu">do.call</span>(<span class="st">'rbind'</span>, .) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(mtdt) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'assay'</span>, <span class="st">'context'</span>)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>data.table<span class="sc">::</span><span class="fu">fwrite</span>(mtdt, <span class="at">file =</span> <span class="fu">file.path</span>(data_dir, <span class="st">'metadata'</span>, <span class="st">'enpact_models_to_train.tsv'</span>), <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>, <span class="at">row.names =</span>F, <span class="at">col.names =</span>T, <span class="at">quote =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>These files created and saved are used to train the DL-based Enpact models.</p>
<p>Steps required to train Enpact models are <a href="https://github.com/hakyimlab/TFPred-snakemake">here</a>. There is a minimal example to follow. As well as the necessary scripts to run the training for AR, FOXA1, HOXB13, and GATA2 in prostate tissues.</p>
</section>
</section>
<section id="train-dl-based-enpact-models-using-prepared-matrix" class="level1">
<h1>Train DL-based Enpact models using prepared matrix</h1>
<p>You may also train the AR-Prostate Enpact model, as well as the others, using the prepared train and test data here: – Train AR_Prostate model</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>train_script <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'src'</span>, <span class="st">'trainEnpactModel.R'</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>evaluate_script <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'src'</span>, <span class="st">'evaluateEnpactModel.R'</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>train_data <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'data'</span>, <span class="st">'enpact/training'</span>, <span class="st">'AR_Prostate.train_epigenome.csv.gz'</span>)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>eval_data <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'data'</span>, <span class="st">'enpact/training'</span>, <span class="st">'AR_Prostate.test_epigenome.csv.gz'</span>)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>model_rds <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'models'</span>, <span class="st">'enpact'</span>, <span class="st">'AR_Prostate_logistic.enpact.rds'</span>)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>basename_eval_output <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'data'</span>, <span class="st">'enpact/evaluations'</span>, <span class="st">'AR_Prostate_logistic.enpact.eval_output'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>cmd <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'sbatch {data_dir}/src/trainAndEvaluateEnpactModel.sbatch {train_script} {evaluate_script} {train_data} {eval_data} {model_rds} {basename_eval_output}'</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>cmd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(cmd) ; <span class="fu">system</span>(<span class="st">'squeue -u temi'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>– Train FOXA1_Prostate model</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>train_script <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'src'</span>, <span class="st">'trainEnpactModel.R'</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>evaluate_script <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'src'</span>, <span class="st">'evaluateEnpactModel.R'</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>train_data <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'data'</span>, <span class="st">'enpact/training'</span>, <span class="st">'FOXA1_Prostate.train_epigenome.csv.gz'</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>eval_data <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'data'</span>, <span class="st">'enpact/training'</span>, <span class="st">'FOXA1_Prostate.test_epigenome.csv.gz'</span>)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>model_rds <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'models'</span>, <span class="st">'enpact'</span>, <span class="st">'FOXA1_Prostate_logistic.enpact.rds'</span>)</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>basename_eval_output <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'data'</span>, <span class="st">'enpact/evaluations'</span>, <span class="st">'FOXA1_Prostate_logistic.enpact.eval_output'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>cmd <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'sbatch {data_dir}/src/trainAndEvaluateEnpactModel.sbatch {train_script} {evaluate_script} {train_data} {eval_data} {model_rds} {basename_eval_output}'</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>cmd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(cmd) ; <span class="fu">system</span>(<span class="st">'squeue -u temi'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>– Train GATA2_Prostate model</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>train_script <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'src'</span>, <span class="st">'trainEnpactModel.R'</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>evaluate_script <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'src'</span>, <span class="st">'evaluateEnpactModel.R'</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>train_data <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'data'</span>, <span class="st">'enpact/training'</span>, <span class="st">'GATA2_Prostate.train_epigenome.csv.gz'</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>eval_data <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'data'</span>, <span class="st">'enpact/training'</span>, <span class="st">'GATA2_Prostate.test_epigenome.csv.gz'</span>)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>model_rds <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'models'</span>, <span class="st">'enpact'</span>, <span class="st">'GATA2_Prostate_logistic.enpact.rds'</span>)</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>basename_eval_output <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'data'</span>, <span class="st">'enpact/evaluations'</span>, <span class="st">'GATA2_Prostate_logistic.enpact.eval_output'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>cmd <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'sbatch {data_dir}/src/trainAndEvaluateEnpactModel.sbatch {train_script} {evaluate_script} {train_data} {eval_data} {model_rds} {basename_eval_output}'</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>cmd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(cmd) ; <span class="fu">system</span>(<span class="st">'squeue -u temi'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>– Train HOXB13_Prostate model</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>train_script <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'src'</span>, <span class="st">'trainEnpactModel.R'</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>evaluate_script <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'src'</span>, <span class="st">'evaluateEnpactModel.R'</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>train_data <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'data'</span>, <span class="st">'enpact/training'</span>, <span class="st">'HOXB13_Prostate.train_epigenome.csv.gz'</span>)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>eval_data <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'data'</span>, <span class="st">'enpact/training'</span>, <span class="st">'HOXB13_Prostate.test_epigenome.csv.gz'</span>)</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>model_rds <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'models'</span>, <span class="st">'enpact'</span>, <span class="st">'HOXB13_Prostate_logistic.enpact.rds'</span>)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>basename_eval_output <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'data'</span>, <span class="st">'enpact/evaluations'</span>, <span class="st">'HOXB13_Prostate_logistic.enpact.eval_output'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>cmd <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'sbatch {data_dir}/src/trainAndEvaluateEnpactModel.sbatch {train_script} {evaluate_script} {train_data} {eval_data} {model_rds} {basename_eval_output}'</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>cmd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(cmd) ; <span class="fu">system</span>(<span class="st">'squeue -u temi'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="prepare-bacas-cwas-models-in-sqlite-and-create-the-necessary-databases" class="level1">
<h1>Prepare Baca’s CWAS models in sqlite and create the necessary databases</h1>
<p>– the CWAS weights for AR</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>transcription_factor <span class="ot">&lt;-</span> <span class="st">'AR'</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>ar_zip <span class="ot">&lt;-</span> <span class="st">'/project2/haky/Data/baca_cwas/cwas_weights/AR.zip'</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">file.exists</span>(ar_zip))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>– First unzip the file</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="fu">glue</span>(<span class="st">'{output_dir}/{transcription_factor}'</span>))){</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>    file_names <span class="ot">&lt;-</span> <span class="fu">unzip</span>(ar_zip, <span class="at">list=</span>T)<span class="sc">$</span>Name</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>    files_to_read <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="at">pattern=</span><span class="st">'^</span><span class="sc">\\</span><span class="st">bAR</span><span class="sc">\\</span><span class="st">b.*</span><span class="sc">\\</span><span class="st">bRDat</span><span class="sc">\\</span><span class="st">b$'</span>, <span class="at">x=</span>file_names, <span class="at">value=</span>T)</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>    files_to_read[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># unzip the file</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>    zip<span class="sc">::</span><span class="fu">unzip</span>(ar_zip, <span class="at">files=</span>files_to_read, <span class="at">exdir=</span>output_dir)</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>} </span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>ar_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="fu">glue</span>(<span class="st">'{output_dir}/{transcription_factor}'</span>))</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>ar_files_locus <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">strsplit</span>(<span class="at">x=</span>ar_files, <span class="at">split=</span><span class="st">'</span><span class="sc">\\</span><span class="st">.'</span>), getElement, <span class="dv">1</span>)</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>ar_files_locus[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>– read weights; Next, read the files <code>.wgt</code> files</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map</span>(<span class="at">.x=</span><span class="fu">seq_along</span>(ar_files_locus), <span class="at">.f=</span><span class="cf">function</span>(i){</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>    locus <span class="ot">&lt;-</span> ar_files_locus[i]</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>    <span class="co">#print(file.exists(glue('{output_dir}/{transcription_factor}/{locus}.wgt.RDat')))</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>    rdt <span class="ot">&lt;-</span> <span class="fu">new.env</span>(<span class="at">parent =</span> <span class="fu">emptyenv</span>())</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">load</span>(<span class="fu">glue</span>(<span class="st">'{output_dir}/{transcription_factor}/{locus}.wgt.RDat'</span>), <span class="at">envir=</span>rdt)</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>    wgts <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(rdt<span class="sc">$</span>wgt.matrix) <span class="sc">%&gt;%</span> </span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>        tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">'snp_id'</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">locus=</span>locus)</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>    snp_info <span class="ot">&lt;-</span> rdt<span class="sc">$</span>snps <span class="sc">%&gt;%</span> </span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">all_of</span>(<span class="fu">c</span>(<span class="st">'V1'</span>, <span class="st">'V3'</span>, <span class="st">'V2'</span>, <span class="st">'V4'</span>, <span class="st">'V5'</span>))) </span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(snp_info) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'chr'</span>, <span class="st">'snp_id'</span>, <span class="st">'position'</span>, <span class="st">'a1'</span>, <span class="st">'a2'</span>)</span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a>    dt <span class="ot">&lt;-</span> base<span class="sc">::</span><span class="fu">merge</span>(wgts, snp_info, <span class="at">by=</span><span class="st">'snp_id'</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">relocate</span>(<span class="fu">all_of</span>(<span class="fu">c</span>(<span class="st">'locus'</span>, <span class="st">'chr'</span>, <span class="st">'position'</span>, <span class="st">'a1'</span>, <span class="st">'a2'</span>)), <span class="at">.after=</span>snp_id)</span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(dt)</span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a>}, <span class="at">.progress=</span>T)</span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a>cwas_db <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">'rbind'</span>, out)</span>
<span id="cb26-22"><a href="#cb26-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-23"><a href="#cb26-23" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(cwas_db) ; cwas_db[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>– write out the weights</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>data.table<span class="sc">::</span><span class="fu">fwrite</span>(cwas_db, <span class="at">file=</span><span class="fu">glue</span>(<span class="st">'{files_dir}/{transcription_factor}_baca_cwas_weights.hg19.{tdate}.txt.gz'</span>), <span class="at">col.names=</span>T, <span class="at">row.names=</span>F, <span class="at">quote=</span>F, <span class="at">compress=</span><span class="st">'gzip'</span>,<span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>cwas_db <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="fu">glue</span>(<span class="st">'{files_dir}/{transcription_factor}_baca_cwas_weights.hg19.{tdate}.txt.gz'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>– read and write out the extras</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map</span>(<span class="at">.x=</span><span class="fu">seq_along</span>(ar_files_locus), <span class="at">.f=</span><span class="cf">function</span>(i){</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>    locus <span class="ot">&lt;-</span> ar_files_locus[i]</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>    <span class="co">#print(file.exists(glue('{output_dir}/{transcription_factor}/{locus}.wgt.RDat')))</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>    rdt <span class="ot">&lt;-</span> <span class="fu">new.env</span>(<span class="at">parent =</span> <span class="fu">emptyenv</span>())</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">load</span>(<span class="fu">glue</span>(<span class="st">'{output_dir}/{transcription_factor}/{locus}.wgt.RDat'</span>), <span class="at">envir=</span>rdt)</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>    cv_perf <span class="ot">&lt;-</span> <span class="fu">rbind</span>(rdt<span class="sc">$</span>cv.performance[<span class="st">'pval'</span>, ], rdt<span class="sc">$</span>cv.performance[<span class="st">'rsq'</span>, ]) <span class="sc">%&gt;%</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">measure =</span> <span class="fu">c</span>(<span class="st">'pval'</span>, <span class="st">'rsq'</span>), locus) <span class="sc">%&gt;%</span></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">relocate</span>(locus, measure) <span class="sc">%&gt;%</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">locus =</span> locus, <span class="at">transcription_factor =</span> transcription_factor, <span class="at">n_snps_in_window =</span> rdt<span class="sc">$</span>N.tot, <span class="at">n.snps.in.model =</span> rdt<span class="sc">$</span>N.as)</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(cv_perf)</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>}, <span class="at">.progress=</span>T)</span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>cvperf_dt <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">'rbind'</span>, out)</span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(cvperf_dt) ; cvperf_dt[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>data.table<span class="sc">::</span><span class="fu">fwrite</span>(cvperf_dt, <span class="at">file=</span><span class="fu">glue</span>(<span class="st">'{files_dir}/{transcription_factor}_baca_cwas_extras.hg19.{tdate}.txt.gz'</span>), <span class="at">col.names=</span>T, <span class="at">row.names=</span>F, <span class="at">quote=</span>F, <span class="at">compress=</span><span class="st">'gzip'</span>,<span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>– map the loci to hg38 from hg19</p>
<section id="bed-mappings-i.e.-liftover-the-arbs-coordinates-from-hg19-to-hg38" class="level3">
<h3 class="anchored" data-anchor-id="bed-mappings-i.e.-liftover-the-arbs-coordinates-from-hg19-to-hg38">Bed mappings i.e.&nbsp;liftover the arbs’ coordinates from hg19 to hg38</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co">#write out the bed files</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>cvperf_dt <span class="sc">%&gt;%</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">separate_wider_delim</span>(<span class="at">col =</span> locus, <span class="at">names =</span> <span class="fu">c</span>(<span class="st">'chr'</span>, <span class="st">'start'</span>, <span class="st">'end'</span>), <span class="at">delim =</span> stringr<span class="sc">::</span><span class="fu">regex</span>(<span class="st">':|-'</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(start, end), as.numeric)) <span class="sc">%&gt;%</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(chr, start, end) <span class="sc">%&gt;%</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">distinct</span>() <span class="sc">%&gt;%</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">id =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(.)) <span class="sc">%&gt;%</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>    data.table<span class="sc">::</span><span class="fu">fwrite</span>(<span class="at">file=</span><span class="fu">glue</span>(<span class="st">'{data_dir}/files/baca_cwas_loci_hg19.bed'</span>), <span class="at">col.names=</span>F, <span class="at">row.names=</span>F, <span class="at">quote=</span>F, <span class="at">sep=</span><span class="st">'</span><span class="sc">\t</span><span class="st">'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># lift over these files</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="co"># download the liftover command : https://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/liftOver</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(<span class="fu">file.path</span>(data_dir, <span class="st">'software'</span>, <span class="st">'liftOver'</span>))){</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">download.file</span>(<span class="st">'https://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/liftOver'</span>, <span class="at">destfile =</span> <span class="fu">file.path</span>(data_dir, <span class="st">'software'</span>, <span class="st">'liftOver'</span>))</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>liftover_sh <span class="ot">&lt;-</span>  <span class="fu">file.path</span>(data_dir, <span class="st">'src'</span>, <span class="st">'liftoverSingleBed.sbatch'</span>)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>liftover_exe <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'software'</span>, <span class="st">'liftOver'</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>input_bed <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'{data_dir}/files/baca_cwas_loci_hg19.bed'</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>chain_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'helpers'</span>, <span class="st">'hg19ToHg38.over.chain.gz'</span>)</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>output_bed <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'{data_dir}/files/baca_cwas_loci_hg38.bed'</span>)</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>unmapped_bed <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'{data_dir}/files/baca_cwas_loci_hg19.unmapped.bed'</span>)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a><span class="fu">file.exists</span>(liftover_sh) ; <span class="fu">file.exists</span>(liftover_exe) ; <span class="fu">file.exists</span>(chain_file) ; <span class="fu">file.exists</span>(input_bed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>cmd <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'sbatch {liftover_sh} {liftover_exe} {chain_file} {input_bed} {output_bed} {unmapped_bed}'</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>cmd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(cmd)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>read in, merge and save</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>hg19_bed <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(input_bed, <span class="at">col.names=</span><span class="fu">c</span>(<span class="st">'chr'</span>, <span class="st">'start.hg19'</span>, <span class="st">'end.19'</span>, <span class="st">'id'</span>))</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>hg38_bed <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(output_bed, <span class="at">col.names=</span><span class="fu">c</span>(<span class="st">'chr'</span>, <span class="st">'start.hg38'</span>, <span class="st">'end.hg39'</span>, <span class="st">'id'</span>))</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>bed_mappings <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">inner_join</span>(hg19_bed, hg38_bed, <span class="at">by=</span><span class="fu">c</span>(<span class="st">'chr'</span> <span class="ot">=</span> <span class="st">'chr'</span>, <span class="st">'id'</span> <span class="ot">=</span> <span class="st">'id'</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>id) <span class="sc">%&gt;%</span> </span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">chrom =</span> chr)</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>data.table<span class="sc">::</span><span class="fu">fwrite</span>(bed_mappings, <span class="at">file=</span><span class="fu">glue</span>(<span class="st">'{data_dir}/files/baca_cwas_arbs_mappings.txt'</span>), <span class="at">col.names=</span>T, <span class="at">row.names=</span>F, <span class="at">quote=</span>F, <span class="at">sep=</span><span class="st">'</span><span class="sc">\t</span><span class="st">'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="snp-mappings-i.e.-liftover-the-snps-coordinates-from-hg19-to-hg38" class="level3">
<h3 class="anchored" data-anchor-id="snp-mappings-i.e.-liftover-the-snps-coordinates-from-hg19-to-hg38">SNP mappings i.e.&nbsp;liftover the snps’ coordinates from hg19 to hg38</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># write out the cwas database snps</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>cwas_db <span class="sc">%&gt;%</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="at">chrom =</span> chr, <span class="at">start =</span> position, locus, <span class="at">rsid=</span>snp_id, a1, a2) <span class="sc">%&gt;%</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">chrom =</span> <span class="fu">paste0</span>(<span class="st">'chr'</span>, chrom, <span class="at">sep=</span><span class="st">''</span>), <span class="at">end =</span> start <span class="sc">+</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">relocate</span>(end, <span class="at">.after =</span> start) <span class="sc">%&gt;%</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>    data.table<span class="sc">::</span><span class="fu">fwrite</span>(<span class="at">file=</span><span class="fu">glue</span>(<span class="st">'{data_dir}/files/baca_cwas_snps_hg19.bed'</span>), <span class="at">col.names=</span>F, <span class="at">row.names=</span>F, <span class="at">quote=</span>F, <span class="at">sep=</span><span class="st">'</span><span class="sc">\t</span><span class="st">'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>liftover_sh <span class="ot">&lt;-</span>  <span class="fu">file.path</span>(data_dir, <span class="st">'src'</span>, <span class="st">'liftoverSingleBed.sbatch'</span>)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>liftover_exe <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'software'</span>, <span class="st">'liftOver'</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>input_bed <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'{data_dir}/files/baca_cwas_snps_hg19.bed'</span>)</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>chain_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'helpers'</span>, <span class="st">'hg19ToHg38.over.chain.gz'</span>)</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>output_bed <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'{data_dir}/files/baca_cwas_snps_hg38.bed'</span>)</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>unmapped_bed <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'{data_dir}/files/baca_cwas_snps_hg19.unmapped.bed'</span>)</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a><span class="fu">file.exists</span>(liftover_sh) ; <span class="fu">file.exists</span>(liftover_exe) ; <span class="fu">file.exists</span>(chain_file) ; <span class="fu">file.exists</span>(input_bed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>cmd <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'sbatch {liftover_sh} {liftover_exe} {chain_file} {input_bed} {output_bed} {unmapped_bed}'</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>cmd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(cmd)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>read in, merge and save</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>hg19_bed <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(input_bed, <span class="at">col.names=</span><span class="fu">c</span>(<span class="st">'chrom'</span>, <span class="st">'start.hg19'</span>, <span class="st">'end.19'</span>, <span class="st">'arbs.hg19'</span>, <span class="st">'rsid'</span>, <span class="st">'a1'</span>, <span class="st">'a2'</span>))</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>hg38_bed <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(output_bed, <span class="at">col.names=</span><span class="fu">c</span>(<span class="st">'chr'</span>, <span class="st">'start.hg38'</span>, <span class="st">'end.hg39'</span>, <span class="st">'arbs.hg19'</span>, <span class="st">'rsid'</span>, <span class="st">'a1'</span>, <span class="st">'a2'</span>))</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>snp_mappings <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">inner_join</span>(hg19_bed, hg38_bed, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">'arbs.hg19'</span> <span class="ot">=</span> <span class="st">'arbs.hg19'</span>, <span class="st">'rsid'</span> <span class="ot">=</span> <span class="st">'rsid'</span>, <span class="st">'a1'</span> <span class="ot">=</span> <span class="st">'a1'</span>, <span class="st">'a2'</span> <span class="ot">=</span> <span class="st">'a2'</span>))</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>data.table<span class="sc">::</span><span class="fu">fwrite</span>(snp_mappings, <span class="at">file=</span><span class="fu">glue</span>(<span class="st">'{data_dir}/files/baca_cwas_snp_mappings.txt'</span>), <span class="at">col.names=</span>T, <span class="at">row.names=</span>F, <span class="at">quote=</span>F, <span class="at">sep=</span><span class="st">'</span><span class="sc">\t</span><span class="st">'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>– Now, you can write out the db and save</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>baca_models <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'lasso'</span>, <span class="st">'lasso.as'</span>, <span class="st">'lasso.plasma'</span>, <span class="st">'top1.as'</span>, <span class="st">'top1.qtl'</span>, <span class="st">'top1'</span>)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>db_folder <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'{data_dir}/models/cwas/db_folder'</span>)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(db_folder)){<span class="fu">dir.create</span>(db_folder)}</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>db_folder_chr <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'{data_dir}/models/cwas/db_folder_chr'</span>)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(db_folder_chr)){<span class="fu">dir.create</span>(db_folder_chr)}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="you-have-to-save-in-predictdb-format" class="level3">
<h3 class="anchored" data-anchor-id="you-have-to-save-in-predictdb-format">You have to save in predictDB format</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>snp_mappings <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="fu">glue</span>(<span class="st">'{data_dir}/files/baca_cwas_snp_mappings.txt'</span>))</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>weights_dt <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="fu">file.path</span>(data_dir, <span class="st">'files'</span>, <span class="st">'AR_baca_cwas_weights.hg19.2024-04-17.txt.gz'</span>))</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>weights_dt <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">inner_join</span>(weights_dt, snp_mappings, <span class="at">by=</span><span class="fu">c</span>(<span class="st">'locus'</span> <span class="ot">=</span> <span class="st">'arbs.hg19'</span>, <span class="st">'snp_id'</span> <span class="ot">=</span> <span class="st">'rsid'</span>, <span class="st">'a1'</span> <span class="ot">=</span> <span class="st">'a1'</span>, <span class="st">'a2'</span> <span class="ot">=</span> <span class="st">'a2'</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="at">rsid =</span> snp_id, locus, chrom, <span class="at">position =</span> start.hg38, a1, a2, <span class="fu">all_of</span>(baca_models))</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(weights_dt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># next, map the locus to hg38</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>bed_mappings <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="fu">glue</span>(<span class="st">'{data_dir}/files/baca_cwas_arbs_mappings.txt'</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">locus.hg19 =</span> <span class="fu">paste0</span>(chrom, <span class="st">':'</span>, start.hg19, <span class="st">'-'</span>, end<span class="fl">.19</span>),</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">locus.hg38 =</span> <span class="fu">paste0</span>(chrom, <span class="st">'_'</span>, start.hg38, <span class="st">'_'</span>, end.hg39)) <span class="sc">%&gt;%</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">select</span>(locus.hg19, locus.hg38)</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>weights_dt <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">inner_join</span>(weights_dt, bed_mappings, <span class="at">by=</span><span class="fu">c</span>(<span class="st">'locus'</span> <span class="ot">=</span> <span class="st">'locus.hg19'</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>locus) <span class="sc">%&gt;%</span></span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">locus =</span> locus.hg38) <span class="sc">%&gt;%</span></span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">relocate</span>(locus, <span class="at">.before =</span> rsid)</span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(weights_dt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read in and map the extras</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>baca_extra <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="fu">glue</span>(<span class="st">'{files_dir}/{transcription_factor}_baca_cwas_extras.hg19.{tdate}.txt.gz'</span>))</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>baca_extra <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">inner_join</span>(baca_extra, bed_mappings, <span class="at">by=</span><span class="fu">c</span>(<span class="st">'locus'</span> <span class="ot">=</span> <span class="st">'locus.hg19'</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>locus) <span class="sc">%&gt;%</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">locus =</span> locus.hg38) <span class="sc">%&gt;%</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">relocate</span>(locus, <span class="at">.before =</span> measure)</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>baca_extra<span class="sc">$</span>pred.perf.qval <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>baca_extra[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>baca_extra <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(locus, measure, transcription_factor, n_snps_in_window, n.snps.in.model, pred.perf.qval, <span class="fu">as.symbol</span>(<span class="st">'lasso'</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">names_from =</span> <span class="st">'measure'</span>, <span class="at">values_from =</span> <span class="st">'lasso'</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">relocate</span>(<span class="fu">c</span>(pval, rsq), <span class="at">.after =</span> locus)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>weights_dt <span class="sc">%&gt;%</span> </span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">varID =</span> <span class="fu">paste0</span>(chrom, <span class="st">'_'</span>, position, <span class="st">'_'</span>, a1, <span class="st">'_'</span>, a2, <span class="at">sep=</span><span class="st">''</span>), <span class="at">gene =</span> locus) <span class="sc">%&gt;%</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(gene, rsid, varID, <span class="at">ref_allele=</span>a1, <span class="at">eff_allele=</span>a2, <span class="at">weight=</span><span class="fu">all_of</span>(<span class="st">'lasso'</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">varID =</span> <span class="fu">gsub</span>(<span class="st">"chr"</span>, <span class="st">''</span>, varID))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>baca_weights_list <span class="ot">&lt;-</span> purrr<span class="sc">::</span><span class="fu">map</span>(<span class="at">.x=</span>baca_models, <span class="cf">function</span>(each_m){</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>    model_weights <span class="ot">&lt;-</span> weights_dt <span class="sc">%&gt;%</span> </span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">varID =</span> <span class="fu">paste0</span>(chrom, <span class="st">'_'</span>, position, <span class="st">'_'</span>, a1, <span class="st">'_'</span>, a2, <span class="at">sep=</span><span class="st">''</span>), <span class="at">gene =</span> locus) <span class="sc">%&gt;%</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">select</span>(gene, rsid, varID, <span class="at">ref_allele=</span>a1, <span class="at">eff_allele=</span>a2, <span class="at">weight=</span><span class="fu">all_of</span>(each_m)) <span class="sc">%&gt;%</span> </span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">varID =</span> <span class="fu">gsub</span>(<span class="st">"chr"</span>, <span class="st">''</span>, varID))</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>    <span class="do">## hg38</span></span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>    edt <span class="ot">&lt;-</span> baca_extra <span class="sc">%&gt;%</span> </span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">select</span>(locus, measure, transcription_factor, n_snps_in_window, n.snps.in.model, pred.perf.qval, <span class="fu">all_of</span>(each_m)) <span class="sc">%&gt;%</span> </span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>        tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">names_from =</span> <span class="st">'measure'</span>, <span class="at">values_from =</span> each_m) <span class="sc">%&gt;%</span> </span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">relocate</span>(<span class="fu">c</span>(pval, rsq), <span class="at">.after =</span> locus) <span class="sc">%&gt;%</span></span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">filter</span>(locus <span class="sc">%in%</span> weights_dt<span class="sc">$</span>locus) <span class="sc">%&gt;%</span></span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>        dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">gene =</span> locus, <span class="at">genename =</span> transcription_factor, <span class="at">pred.perf.R2 =</span> rsq, <span class="at">pred.perf.pval =</span> pval)</span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a>    each_db <span class="ot">&lt;-</span> DBI<span class="sc">::</span><span class="fu">dbConnect</span>(RSQLite<span class="sc">::</span><span class="fu">SQLite</span>(), <span class="fu">glue</span>(<span class="st">'{db_folder}/baca_cwas_{each_m}.hg38.db'</span>))</span>
<span id="cb47-16"><a href="#cb47-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dbWriteTable</span>(each_db, <span class="st">"extra"</span>, edt, <span class="at">overwrite=</span>T)</span>
<span id="cb47-17"><a href="#cb47-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dbWriteTable</span>(each_db, <span class="st">"weights"</span>, model_weights, <span class="at">overwrite=</span>T)</span>
<span id="cb47-18"><a href="#cb47-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dbDisconnect</span>(each_db)</span>
<span id="cb47-19"><a href="#cb47-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-20"><a href="#cb47-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="dv">0</span>)</span>
<span id="cb47-21"><a href="#cb47-21" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb47-22"><a href="#cb47-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-23"><a href="#cb47-23" aria-hidden="true" tabindex="-1"></a><span class="co"># names(baca_weights_list) &lt;- baca_models</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>– see an example</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># '/project2/haky/temi/projects/TFXcan/baca_cwas/db_folder/baca_cwas_lasso.db'</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>mydb <span class="ot">&lt;-</span> <span class="fu">dbConnect</span>(<span class="fu">SQLite</span>(), <span class="fu">glue</span>(<span class="st">'{db_folder}/baca_cwas_lasso.hg38.db'</span>))</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>ex <span class="ot">&lt;-</span> <span class="fu">dbGetQuery</span>(mydb, <span class="st">'SELECT * FROM extra'</span>)</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>wt <span class="ot">&lt;-</span> <span class="fu">dbGetQuery</span>(mydb, <span class="st">'SELECT * FROM weights'</span>)</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>ex <span class="sc">|&gt;</span> <span class="fu">head</span>(); wt <span class="sc">|&gt;</span> <span class="fu">head</span>()</span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dbDisconnect</span>(mydb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="predict-cwas-scores-using-these-models" class="level1">
<h1>Predict CWAS scores using these models</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>db_folder <span class="ot">&lt;-</span> <span class="st">'/project/haky/users/temi/projects/Enpact/models/cwas/db_folder'</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>txt_genotypes <span class="ot">&lt;-</span> <span class="st">'/project2/haky/Data/baca_cwas/vcfs/hg38/formatted_geno/all_chrs.text_dosages.txt.gz'</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>txt_samples <span class="ot">&lt;-</span> <span class="st">'/project2/haky/Data/baca_cwas/vcfs/hg38/formatted_geno/samples.text_dosages.txt'</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>exec_file <span class="ot">&lt;-</span> <span class="st">'/beagle3/haky/users/temi/software/MetaXcan/software/Predict.py'</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>output_folder <span class="ot">&lt;-</span> <span class="st">"/project/haky/users/temi/projects/Enpact/data/baca_cwas/output"</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(output_folder)){<span class="fu">dir.create</span>(output_folder, <span class="at">recursive =</span> T)}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># if needed, you should edit the sbatch file: {data_dir}/src/predictCWASscores.sbatch</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>cmd <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'sbatch {data_dir}/src/predictCWASscores.sbatch {db_folder} {txt_genotypes} {txt_samples} {exec_file} {output_folder}'</span>)</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>cmd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(cmd) ; <span class="fu">system</span>(<span class="st">'squeue -u temi'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>prepare the file in a simple matrix</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read in the individual mappings</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>samples_metadata <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="fu">file.path</span>(data_dir, <span class="st">'metadata'</span>, <span class="st">'baca_samples_mappings.metadata.txt'</span>))</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>cwas_scores <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="fu">glue</span>(<span class="st">'{output_folder}/top1.hg38/baca_cwas_predict.txt'</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>FID) <span class="sc">%&gt;%</span></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="st">'IID'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">t</span>() <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>cwas_mat <span class="ot">&lt;-</span> cwas_scores[,]</span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(cwas_mat) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">'UW_PDX_172'</span>, <span class="st">'UW_PDX_170_2'</span>, <span class="fu">colnames</span>(cwas_mat))</span>
<span id="cb52-12"><a href="#cb52-12" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(cwas_mat) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">'UW_PDX_173'</span>, <span class="st">'UW_PDX_170_3'</span>, <span class="fu">colnames</span>(cwas_mat))</span>
<span id="cb52-13"><a href="#cb52-13" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(cwas_mat) <span class="ot">&lt;-</span> samples_metadata[<span class="fu">match</span>(<span class="fu">colnames</span>(cwas_mat), samples_metadata<span class="sc">$</span>vcf), ]<span class="sc">$</span>id</span>
<span id="cb52-14"><a href="#cb52-14" aria-hidden="true" tabindex="-1"></a>cwas_mat <span class="ot">&lt;-</span> cwas_mat <span class="sc">%&gt;%</span> tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">'locus'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>data.table<span class="sc">::</span><span class="fu">fwrite</span>(cwas_mat, <span class="at">file=</span><span class="fu">glue</span>(<span class="st">'{data_dir}/files/baca_cwas_scores.hg38.txt.gz'</span>), <span class="at">col.names=</span>T, <span class="at">row.names=</span>F, <span class="at">quote=</span>F, <span class="at">compress=</span><span class="st">'gzip'</span>, <span class="at">sep=</span><span class="st">'</span><span class="sc">\t</span><span class="st">'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="predict-enpact-scores-using-the-dl-based-enpact-model" class="level1">
<h1>Predict Enpact scores using the DL-based Enpact model</h1>
<p>The steps here are involved.</p>
<ol type="1">
<li>Get epigenomic features for all Baca individuals with Enformer. In this paper, we used <a href="">this pipeline</a>.</li>
<li>With the DL-based Enpact model, predict the Enpact scores using <a href="https://github.com/hakyimlab/enpact-predict-snakemake">this pipeline</a></li>
</ol>
</section>
<section id="train-a-snp-based-enpact-model" class="level1">
<h1>Train a SNP-based Enpact model</h1>
<ol type="1">
<li>Get epigenomic features across 521 EUR individuals with Enformer; In this paper, we used <a href="">this pipeline</a>.</li>
<li>With the DL-based Enpact model, predict the Enpact scores using <a href="https://github.com/hakyimlab/enpact-predict-snakemake">this pipeline</a></li>
<li>Prepare the matrix and data for the SNP-based Enpact model such that they are compatible with predictDB format</li>
<li>Train the SNP-based Enpact model using <a href="">this pipeline</a></li>
</ol>
</section>
<section id="predict-enpact-scores-using-the-snp-based-enpact-model" class="level1">
<h1>Predict Enpact scores using the SNP-based Enpact model</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>db_name <span class="ot">&lt;-</span> <span class="st">'EUR_AR_Prostate_logistic'</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>model_db <span class="ot">&lt;-</span> <span class="st">'/project/haky/users/temi/projects/Enpact/models/lenpact/predict_db_EUR_AR_Prostate_logistic_filtered.db'</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>txt_genotypes <span class="ot">&lt;-</span> <span class="st">'/project2/haky/Data/baca_cwas/vcfs/hg38/formatted_geno/all_chrs.text_dosages.txt.gz'</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>txt_samples <span class="ot">&lt;-</span> <span class="st">'/project2/haky/Data/baca_cwas/vcfs/hg38/formatted_geno/samples.text_dosages.txt'</span></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>exec_file <span class="ot">&lt;-</span> <span class="st">'/beagle3/haky/users/temi/software/MetaXcan/software/Predict.py'</span></span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>output_folder <span class="ot">&lt;-</span> <span class="st">"/project/haky/users/temi/projects/Enpact/data/baca_lenpact/output"</span></span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(output_folder)){<span class="fu">dir.create</span>(output_folder, <span class="at">recursive =</span> T)}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># if needed, you should edit the sbatch file: {data_dir}/src/predictCWASscores.sbatch</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>cmd <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'sbatch {data_dir}/src/predictENPACTscores.sbatch {db_name} {model_db} {txt_genotypes} {txt_samples} {exec_file} {output_folder}'</span>)</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>cmd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(cmd) ; <span class="fu">system</span>(<span class="st">'squeue -u temi'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="running-tfxcan-on-prostate-cancer-data" class="level1">
<h1>Running TFXcan on prostate cancer data</h1>
<p>You can download the GWAS summary statistic from the CWAS paper repo <a href="https://github.com/scbaca/cwas/blob/master/gwas_data/ProstateCancer_Meta_Schumacher2018.nodup.sumstats.gz">here</a> or, better still, from the <a href="https://www.ebi.ac.uk/gwas/publications/29892016">GWAS catalog</a>. This is <a href="https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST006001-GCST007000/GCST006085/harmonised/29892016-GCST006085-EFO_0001663.h.tsv.gz">the direct link</a></p>
<p>– If you have/want to download from the GWAS catalog, this following scripts should help with processing</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>gwas_data <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'data'</span>, <span class="st">'sumstats'</span>, <span class="st">'29892016-GCST006085-EFO_0001663.h.tsv.gz'</span>)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">dirname</span>(gwas_data), <span class="at">recursive =</span> F)</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(gwas_data)){</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">download.file</span>(<span class="st">'https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST006001-GCST007000/GCST006085/harmonised/29892016-GCST006085-EFO_0001663.h.tsv.gz'</span>, <span class="at">destfile =</span> gwas_data)</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>– process the downloaded file. I have processed this file in accordance with the weights data from the SNP-based Enpact model. So, you will need to read in the models and use the data to filter for compatible SNPs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>pcr_sumstat <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(gwas_data)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(pcr_sumstat); pcr_sumstat[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>– read in the lEnpact weights</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># model_db &lt;- file.path(data_dir, 'models', 'lenpact', 'predict_db_EUR_AR_Prostate_logistic_filtered.db')</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>enpactdb <span class="ot">&lt;-</span> <span class="fu">dbConnect</span>(<span class="fu">SQLite</span>(), model_db)</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>weights <span class="ot">&lt;-</span> <span class="fu">dbGetQuery</span>(enpactdb, <span class="st">'SELECT * FROM weights'</span>)</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dbDisconnect</span>(enpactdb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>reformat the file and filter for only those with variant Id in the weights data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>pcgss <span class="ot">&lt;-</span> pcr_sumstat <span class="sc">%&gt;%</span> </span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="at">chrom=</span>hm_chrom, <span class="at">variant_id=</span>hm_variant_id, <span class="at">rsid=</span>hm_rsid, <span class="at">pos=</span>hm_pos, <span class="at">A2=</span>hm_other_allele, <span class="at">A1=</span>hm_effect_allele, <span class="at">beta=</span>hm_beta, <span class="at">p_value=</span>p_value, <span class="at">se=</span>standard_error, <span class="at">maf=</span>hm_effect_allele_frequency) <span class="sc">%&gt;%</span> </span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(<span class="sc">!</span>(<span class="fu">is.na</span>(variant_id) <span class="sc">|</span> <span class="fu">is.na</span>(rsid) <span class="sc">|</span> <span class="fu">is.na</span>(pos))) <span class="sc">%&gt;%</span></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">zscore=</span>beta<span class="sc">/</span>se) <span class="sc">%&gt;%</span> </span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(variant_id <span class="sc">%in%</span> weights<span class="sc">$</span>varID)</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>pcgss[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]; <span class="fu">dim</span>(pcgss)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>output_folder <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'data'</span>, <span class="st">'sumstats'</span>)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(output_folder)){</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dir.create</span>(output_folder, <span class="at">recursive=</span>T)</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>pcgss <span class="sc">%&gt;%</span> <span class="fu">split</span>(.<span class="sc">$</span>chrom) <span class="sc">%&gt;%</span> <span class="fu">imap</span>(<span class="sc">~</span>data.table<span class="sc">::</span><span class="fu">fwrite</span>(.x, <span class="fu">glue</span>(<span class="st">'{output_folder}/chr{.y}_Schumacher.gwas_ss.txt.gz'</span>), <span class="at">compress=</span><span class="st">'gzip'</span>, <span class="at">row.names=</span>F, <span class="at">quote=</span>F, <span class="at">sep =</span> <span class="st">'</span><span class="sc">\t</span><span class="st">'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>– Now, you can run TFXcan on the prostate cancer data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>db_name <span class="ot">&lt;-</span> <span class="st">'EUR_AR_Prostate_logistic'</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>model_db <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'models/lenpact/predict_db_EUR_AR_Prostate_logistic_filtered.db'</span>)</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>covariances <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'models/lenpact/Covariances.varID.txt'</span>)</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>gwas_folder <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'data/sumstats'</span>)</span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>gwas_file_pattern <span class="ot">&lt;-</span> <span class="st">".*_Schumacher.gwas_ss.txt.gz"</span></span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a>exec_file <span class="ot">&lt;-</span> <span class="st">'/beagle3/haky/users/temi/software/MetaXcan/software/SPrediXcan.py'</span></span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a>output_file <span class="ot">&lt;-</span> <span class="fu">file.path</span>(data_dir, <span class="st">'data'</span>, <span class="st">'tfxcan'</span>, <span class="st">'EUR_AR_Prostate_logistic.TFXcan.prostate_cancer_risk.csv'</span>)</span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">dir.exists</span>(<span class="fu">dirname</span>(output_file))){<span class="fu">dir.create</span>(<span class="fu">dirname</span>(output_file), <span class="at">recursive =</span> T)}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>cmd <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'sbatch {data_dir}/src/sTFXcan.sbatch {db_name} {model_db} {covariances} {gwas_folder} {gwas_file_pattern} {exec_file} {output_file}'</span>)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>cmd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(cmd) ; <span class="fu">system</span>(<span class="st">'squeue -u temi'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="running-twas-on-prostate-cancer-data" class="level1">
<h1>Running TWAS on prostate cancer data</h1>
<p>This is very similar to using the PrediXcan pipeline. You will need to download the TWAS weights from the <a href="https://predictdb.org/">PrediXcan website</a>. You can use the <a href="https://predictdb.org/download/weights/GTEx_V8_HapMap-2017-11-29.tar.gz">GTEx v8 weights</a> or the <a href="https://predictdb.org/download/weights/GTEx_V7_HapMap-2017-11-29.tar.gz">GTEx v7 weights</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">knit_exit</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>