<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.0.38">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Temi">

<title>TF binding analysis | GATA3 tracks only</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="impact-enformer-modelling-3_files/libs/clipboard/clipboard.min.js"></script>
<script src="impact-enformer-modelling-3_files/libs/quarto-html/quarto.js"></script>
<script src="impact-enformer-modelling-3_files/libs/quarto-html/popper.min.js"></script>
<script src="impact-enformer-modelling-3_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="impact-enformer-modelling-3_files/libs/quarto-html/anchor.min.js"></script>
<link href="impact-enformer-modelling-3_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="impact-enformer-modelling-3_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="impact-enformer-modelling-3_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="impact-enformer-modelling-3_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="impact-enformer-modelling-3_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">TF binding analysis | GATA3 tracks only</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Temi </p>
          </div>
  </div>
    
    
  </div>
  

</header>

<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span>opts_chunk<span class="sc">$</span><span class="fu">set</span>(<span class="at">fig.width=</span><span class="dv">12</span>, <span class="at">fig.height=</span><span class="dv">8</span>, <span class="at">cache=</span>F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list=</span><span class="fu">ls</span>())</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">'/projects/covid-ct/imlab/users/temi/projects/TFXcan/scripts/'</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glue)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(R.utils)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glmnet)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(doMC)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ROCR)</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Matrix)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reshape2)</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(foreach)</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(doParallel)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>plots_dir <span class="ot">&lt;-</span> <span class="st">'../plots'</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>models_dir <span class="ot">&lt;-</span> <span class="st">'../models'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>TF <span class="ot">&lt;-</span> <span class="st">'GATA3'</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>cistrome_dir <span class="ot">&lt;-</span> <span class="st">'/projects/covid-ct/imlab/data/cistrome/compressed'</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>hf_info <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="fu">glue</span>(<span class="st">'{cistrome_dir}/human_factor_full_QC.txt'</span>))</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>TF_DCID <span class="ot">&lt;-</span> hf_info[hf_info<span class="sc">$</span>Factor <span class="sc">==</span> TF, ]</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(TF_DCID)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    DCid      Species     GSMID Factor Cell_line       Cell_type Tissue_type
1:  2324 Homo sapiens GSM720423  GATA3     MCF-7      Epithelium      Breast
2:  2325 Homo sapiens GSM720422  GATA3     MCF-7      Epithelium      Breast
3:  9195 Homo sapiens GSM957608  GATA3 RPMI-8402 T cell leukemia       Blood
4:  9196 Homo sapiens GSM957609  GATA3 RPMI-8402 T cell leukemia       Blood
5: 33129 Homo sapiens GSM986070  GATA3     MCF-7      Epithelium      Breast
6: 33136 Homo sapiens GSM986075  GATA3     MCF-7      Epithelium      Breast
   FastQC UniquelyMappedRatio   PBC PeaksFoldChangeAbove10       FRiP
1:     30              0.7981 0.980                   5420 0.03490225
2:     30              0.8010 0.989                  12454 0.08374850
3:     39              0.7523 0.244                   1432 0.08693400
4:     39              0.7604 0.808                   3459 0.04521850
5:     39              0.8001 0.990                   8318 0.04329025
6:     39              0.7460 0.971                    358 0.00357925
   PeaksUnionDHSRatio
1:          0.9750000
2:          0.9810000
3:          0.8848000
4:          0.9524000
5:          0.9618000
6:          0.8949343</code></pre>
</div>
</div>
<p>https://stats.stackexchange.com/questions/138569/why-is-lambda-within-one-standard-error-from-the-minimum-is-a-recommended-valu</p>
<div style="page-break-after: always;"></div>
<p>For plotting, I need to load in the enformer annotations</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>enformer_annotations <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="st">'../info-files/enformer_tracks_annotated-resaved.txt'</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>enformer_annotations <span class="ot">&lt;-</span> enformer_annotations[<span class="sc">!</span><span class="fu">is.na</span>(enformer_annotations<span class="sc">$</span>assay), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># using an individual for now &gt;&gt; will do this across all 3 individuals later</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>ind <span class="ot">&lt;-</span> <span class="st">'HG00479'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There are 5 major categories</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>assays <span class="ot">&lt;-</span> enformer_annotations<span class="sc">$</span>assay <span class="sc">|&gt;</span> <span class="fu">unique</span>()</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>assays</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "DNase-seq"        "ATAC-seq"         "TF ChIP-seq"      "Histone ChIP-seq"
[5] "CAGE experiment" </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>z_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(assays)){</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>    z <span class="ot">&lt;-</span> <span class="fu">combn</span>(assays, i)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    z_list[[i]] <span class="ot">&lt;-</span> <span class="fu">t</span>(z) <span class="sc">|&gt;</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>assays_dt <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">rbind.fill</span>(z_list)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># registerDoMC(cores = 24)</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co"># enet_models_aggByMean &lt;- list()</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co"># # pick the tracks</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co"># for(i in 1:nrow(assays_dt)){</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="co">#     cat('Currently on', i, '\n')</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="co">#     # select the categories</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a><span class="co">#     track_categories &lt;- assays_dt[i, ][!is.na(assays_dt[i, ])]</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="co">#     selected_tracks &lt;- enformer_annotations$feature_names[enformer_annotations$assay %in% track_categories]</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="co">#     X_train_selected &lt;- X_train[, colnames(X_train) %in% selected_tracks]</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a><span class="co">#     X_test_selected &lt;- X_test[, colnames(X_test) %in% selected_tracks]</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a><span class="co">#     enet_fit_selected &lt;- cv.glmnet(x=X_train_selected[complete.cases(X_train_selected),], y=y_train[complete.cases(X_train_selected)], family = "binomial", type.measure = "auc", alpha = 0.5, parallel=T, keep=T)</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a><span class="co">#     enet_models_aggByMean[[i]] &lt;- enet_fit_selected</span></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a><span class="co"># }</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="helper-functions" class="level3">
<h3 class="anchored" data-anchor-id="helper-functions">helper functions</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>build_save_list_model <span class="ot">&lt;-</span> <span class="cf">function</span>(){</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    <span class="co"># this function by default uses whatever X_train and y_train is in the environment</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>    cl <span class="ot">&lt;-</span> <span class="fu">makeCluster</span>(<span class="dv">12</span>)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">registerDoParallel</span>(cl)</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">getDoParName</span>())</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>    assays_dt <span class="ot">&lt;-</span> <span class="fu">get</span>(<span class="st">'assays_dt'</span>, .GlobalEnv)</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>    enformer_annotations <span class="ot">&lt;-</span> <span class="fu">get</span>(<span class="st">'enformer_annotations'</span>, .GlobalEnv)</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>    X_train <span class="ot">&lt;-</span> <span class="fu">get</span>(<span class="st">'X_train'</span>, .GlobalEnv)</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>    y_train <span class="ot">&lt;-</span> <span class="fu">get</span>(<span class="st">'y_train'</span>, .GlobalEnv)</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>    enet_models_list <span class="ot">&lt;-</span> <span class="fu">foreach</span>(<span class="at">i =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(assays_dt)) <span class="sc">%dopar%</span> {</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>        <span class="co">#cat('Currently on', i, '\n')</span></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>        <span class="co"># select the categories</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>        track_categories <span class="ot">&lt;-</span> assays_dt[i, ][<span class="sc">!</span><span class="fu">is.na</span>(assays_dt[i, ])]</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>        selected_tracks <span class="ot">&lt;-</span> enformer_annotations<span class="sc">$</span>feature_names[enformer_annotations<span class="sc">$</span>assay <span class="sc">%in%</span> track_categories]</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>        X_train_selected <span class="ot">&lt;-</span> X_train[, <span class="fu">colnames</span>(X_train) <span class="sc">%in%</span> selected_tracks]</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>        <span class="co">#X_test_selected &lt;- X_test[, colnames(X_test) %in% selected_tracks]</span></span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>        enet_fit_selected <span class="ot">&lt;-</span> glmnet<span class="sc">::</span><span class="fu">cv.glmnet</span>(<span class="at">x=</span>X_train_selected[<span class="fu">complete.cases</span>(X_train_selected),], <span class="at">y=</span>y_train[<span class="fu">complete.cases</span>(X_train_selected)], <span class="at">family =</span> <span class="st">"binomial"</span>, <span class="at">type.measure =</span> <span class="st">"auc"</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">parallel=</span>T, <span class="at">keep=</span>T)</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>        enet_fit_selected</span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stopCluster</span>(cl)</span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(enet_models_list)</span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a>collate_coefficients <span class="ot">&lt;-</span> <span class="cf">function</span>(fit){</span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a>    min_error_index <span class="ot">&lt;-</span> fit<span class="sc">$</span>index[<span class="st">'min'</span>, ]</span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a>    one_sd_index <span class="ot">&lt;-</span> fit<span class="sc">$</span>index[<span class="st">'1se'</span>, ]</span>
<span id="cb12-34"><a href="#cb12-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-35"><a href="#cb12-35" aria-hidden="true" tabindex="-1"></a>    dimensions <span class="ot">&lt;-</span> fit<span class="sc">$</span>glmnet.fit<span class="sc">$</span>beta<span class="sc">@</span>Dim</span>
<span id="cb12-36"><a href="#cb12-36" aria-hidden="true" tabindex="-1"></a>    coef_mat <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">summary</span>(fit<span class="sc">$</span>glmnet.fit<span class="sc">$</span>beta))</span>
<span id="cb12-37"><a href="#cb12-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-38"><a href="#cb12-38" aria-hidden="true" tabindex="-1"></a>    temp_mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data=</span><span class="cn">NA</span>, <span class="at">nrow=</span>dimensions[<span class="dv">1</span>], <span class="at">ncol=</span>dimensions[<span class="dv">2</span>])</span>
<span id="cb12-39"><a href="#cb12-39" aria-hidden="true" tabindex="-1"></a>    <span class="co">#print(dim(temp_mat))</span></span>
<span id="cb12-40"><a href="#cb12-40" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(coef_mat)){</span>
<span id="cb12-41"><a href="#cb12-41" aria-hidden="true" tabindex="-1"></a>        temp_mat[coef_mat[i, <span class="st">'i'</span>], coef_mat[i, <span class="st">'j'</span>]] <span class="ot">&lt;-</span> coef_mat[i, <span class="st">'x'</span>]</span>
<span id="cb12-42"><a href="#cb12-42" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb12-43"><a href="#cb12-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-44"><a href="#cb12-44" aria-hidden="true" tabindex="-1"></a>    temp_mat[<span class="fu">is.na</span>(temp_mat)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb12-45"><a href="#cb12-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-46"><a href="#cb12-46" aria-hidden="true" tabindex="-1"></a>    fit_beta <span class="ot">&lt;-</span> <span class="fu">cbind</span>(temp_mat[, min_error_index], temp_mat[, one_sd_index])</span>
<span id="cb12-47"><a href="#cb12-47" aria-hidden="true" tabindex="-1"></a>    <span class="co"># what features were used?</span></span>
<span id="cb12-48"><a href="#cb12-48" aria-hidden="true" tabindex="-1"></a>    feature_data <span class="ot">&lt;-</span> enformer_annotations[enformer_annotations<span class="sc">$</span>feature_names <span class="sc">%in%</span> (fit<span class="sc">$</span>glmnet.fit<span class="sc">$</span>beta <span class="sc">|&gt;</span> <span class="fu">rownames</span>()), <span class="fu">c</span>(<span class="st">'assay'</span>, <span class="st">'feature_names'</span>)]</span>
<span id="cb12-49"><a href="#cb12-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-50"><a href="#cb12-50" aria-hidden="true" tabindex="-1"></a>    fit_beta <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">cbind</span>(fit_beta, feature_data))</span>
<span id="cb12-51"><a href="#cb12-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-52"><a href="#cb12-52" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(fit_beta)</span>
<span id="cb12-53"><a href="#cb12-53" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb12-54"><a href="#cb12-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-55"><a href="#cb12-55" aria-hidden="true" tabindex="-1"></a>test_models <span class="ot">&lt;-</span> <span class="cf">function</span>(model, X_test_set, y_test_set){</span>
<span id="cb12-56"><a href="#cb12-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-57"><a href="#cb12-57" aria-hidden="true" tabindex="-1"></a>    features <span class="ot">&lt;-</span> model<span class="sc">$</span>glmnet.fit<span class="sc">$</span>beta <span class="sc">|&gt;</span> <span class="fu">rownames</span>()</span>
<span id="cb12-58"><a href="#cb12-58" aria-hidden="true" tabindex="-1"></a>    X_test_set <span class="ot">&lt;-</span> X_test_set[, features]</span>
<span id="cb12-59"><a href="#cb12-59" aria-hidden="true" tabindex="-1"></a>    <span class="fu">assess.glmnet</span>(model, <span class="at">newx =</span> X_test_set, <span class="at">newy =</span> y_test_set) <span class="sc">|&gt;</span> <span class="fu">unlist</span>()</span>
<span id="cb12-60"><a href="#cb12-60" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="models" class="level1">
<h1>Models …</h1>
<p>Here, each predictor is the average of the 8 bins upstream and downstream the bin where the center of the motif resides.</p>
<p>Loading the data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>test_data <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="fu">glue</span>(<span class="st">'/projects/covid-ct/imlab/users/temi/projects/TFXcan/output/train-test-data/test_aggByMean_{ind}_{TF}.csv'</span>))</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>X_test <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(test_data[, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>)])</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>y_test <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(test_data[, <span class="dv">1</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>aggModels_GATA3_names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'aggByCenter'</span>, <span class="st">'aggByMean'</span>, <span class="st">'aggByMeanUpstream'</span>, <span class="st">'aggByMeanDownstream'</span>, <span class="st">'aggByMeanUpstreamDownstream'</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>model_name <span class="ot">&lt;-</span> <span class="fu">glue</span>(<span class="st">'enet_aggModels_{TF}_tracks_only.RData'</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">file.exists</span>(<span class="fu">glue</span>(<span class="st">'{models_dir}/{model_name}'</span>))){</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="st">'Models exists. Reading them in...</span><span class="sc">\n</span><span class="st">'</span>)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    enet_models_GATA3 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">glue</span>(<span class="st">'{models_dir}/{model_name}'</span>))</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    enet_models_GATA3 <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(aggModels_GATA3_names)){</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>        train_data <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="fu">glue</span>(<span class="st">'/projects/covid-ct/imlab/users/temi/projects/TFXcan/output/train-test-data/train_{aggModels_GATA3_names[i]}_{ind}_{TF}.csv'</span>))</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>        X_train <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(train_data[, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>)])</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>        y_train <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(train_data[, <span class="dv">1</span>])</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>        enet_model <span class="ot">&lt;-</span> glmnet<span class="sc">::</span><span class="fu">cv.glmnet</span>(<span class="at">x=</span>X_train[<span class="fu">complete.cases</span>(X_train),], <span class="at">y=</span>y_train[<span class="fu">complete.cases</span>(X_train)], <span class="at">family =</span> <span class="st">"binomial"</span>, <span class="at">type.measure =</span> <span class="st">"auc"</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">parallel=</span>T, <span class="at">keep=</span>T)</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>        enet_models_GATA3[[i]] <span class="ot">&lt;-</span> enet_model</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Models exists. Reading them in...\n"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(enet_models_GATA3) <span class="ot">&lt;-</span> aggModels_GATA3_names</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co">#save the model for next time</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(enet_models_GATA3, <span class="at">file=</span><span class="fu">glue</span>(<span class="st">'{models_dir}/{model_name}'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>fit_beta_labels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'For lambda with minimum mean cv error'</span>, <span class="st">'For largest lambda with 1 s.e within the mininum error'</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="co">#fit &lt;- enet_models_aggByMean[[1]]</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>df_beta_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(enet_models_GATA3, collate_coefficients)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>df_beta_list</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$aggByCenter
          V1        V2       assay feature_names
1 -0.1573064 0.0000000 TF ChIP-seq         f_983
2  0.4707158 0.4051203 TF ChIP-seq        f_1417
3  0.5623215 0.2318938 TF ChIP-seq        f_2740

$aggByMean
         V1        V2       assay feature_names
1 0.0000000 0.0000000 TF ChIP-seq         f_983
2 0.0000000 0.0000000 TF ChIP-seq        f_1417
3 0.0316565 0.0316565 TF ChIP-seq        f_2740

$aggByMeanUpstream
          V1         V2       assay feature_names
1 0.00000000 0.00000000 TF ChIP-seq         f_983
2 0.00000000 0.00000000 TF ChIP-seq        f_1417
3 0.02307838 0.02307838 TF ChIP-seq        f_2740

$aggByMeanDownstream
          V1         V2       assay feature_names
1 0.00000000 0.00000000 TF ChIP-seq         f_983
2 0.00000000 0.00000000 TF ChIP-seq        f_1417
3 0.02525063 0.02525063 TF ChIP-seq        f_2740

$aggByMeanUpstreamDownstream
          V1         V2       assay feature_names
1 0.00000000 0.00000000 TF ChIP-seq         f_983
2 0.00000000 0.00000000 TF ChIP-seq        f_1417
3 0.03195424 0.03195424 TF ChIP-seq        f_2740</code></pre>
</div>
</div>
<p>Performance on the test set</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">lapply</span>(enet_models_GATA3, test_models, X_test, y_test)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, out)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                            deviance.lambda.1se class.lambda.1se      auc
aggByCenter                            1.312376            0.421 0.641184
aggByMean                              1.378715            0.373 0.677704
aggByMeanUpstream                      1.380718            0.399 0.677704
aggByMeanDownstream                    1.380207            0.367 0.677704
aggByMeanUpstreamDownstream            1.378647            0.367 0.677704
                            mse.lambda.1se mae.lambda.1se
aggByCenter                      0.4651309      0.9432517
aggByMean                        0.4962129      0.9960838
aggByMeanUpstream                0.4972127      0.9971440
aggByMeanDownstream              0.4969578      0.9968754
aggByMeanUpstreamDownstream      0.4961787      0.9960470</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>performance_metrics <span class="ot">&lt;-</span> out[<span class="fu">order</span>(out[, <span class="dv">4</span>], <span class="at">decreasing=</span>F), ]</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>performance_metrics</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                            deviance.lambda.1se class.lambda.1se      auc
aggByCenter                            1.312376            0.421 0.641184
aggByMeanUpstreamDownstream            1.378647            0.367 0.677704
aggByMean                              1.378715            0.373 0.677704
aggByMeanDownstream                    1.380207            0.367 0.677704
aggByMeanUpstream                      1.380718            0.399 0.677704
                            mse.lambda.1se mae.lambda.1se
aggByCenter                      0.4651309      0.9432517
aggByMeanUpstreamDownstream      0.4961787      0.9960470
aggByMean                        0.4962129      0.9960838
aggByMeanDownstream              0.4969578      0.9968754
aggByMeanUpstream                0.4972127      0.9971440</code></pre>
</div>
</div>
<p>The best model based on the lowest mse</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>bmodel_index <span class="ot">&lt;-</span> <span class="fu">which.min</span>(out[, <span class="dv">4</span>])</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>bmodel <span class="ot">&lt;-</span> enet_models_GATA3[[bmodel_index]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Some diagnostic plots</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>used_categories <span class="ot">&lt;-</span> <span class="st">'TF (GATA3) ChIP-seq'</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">oma=</span><span class="fu">c</span>(<span class="fl">2.0</span>, <span class="fl">2.0</span>, <span class="dv">3</span>, <span class="fl">2.0</span>) <span class="sc">+</span> <span class="fl">0.1</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>bmodel <span class="sc">|&gt;</span> <span class="fu">plot</span>()</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="fu">glue</span>(<span class="st">'{rownames(out)[bmodel_index]}: {used_categories}'</span>), <span class="at">side=</span><span class="dv">3</span>, <span class="at">line=</span><span class="dv">3</span>, <span class="at">cex=</span><span class="fl">1.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="impact-enformer-modelling-3_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>bmodel_df_beta <span class="ot">&lt;-</span> df_beta_list[[bmodel_index]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>useColors <span class="ot">&lt;-</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="at">n=</span><span class="dv">8</span>, <span class="at">name=</span><span class="st">'Dark2'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>df_beta_labels <span class="ot">&lt;-</span> fit_beta_labels</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>track_names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'GATA3:T47D'</span>, <span class="st">'GATA3:SH-SY5Y'</span>, <span class="st">'GATA3:MCF-7'</span>)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="co"># define the layout</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="fu">layout</span>(<span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">2</span>,<span class="dv">3</span>), <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">byrow =</span> T), <span class="at">height=</span><span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">5</span>), <span class="at">width=</span><span class="fu">c</span>(<span class="dv">8</span>,<span class="dv">1</span>))</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar=</span><span class="fu">c</span>(<span class="fl">0.5</span>,<span class="fl">2.5</span>,<span class="dv">2</span>,<span class="dv">2</span>) <span class="sc">+</span> <span class="fl">0.1</span>, <span class="at">oma=</span><span class="fu">c</span>(<span class="fl">3.5</span>, <span class="fl">2.0</span>, <span class="fl">3.5</span>, <span class="fl">0.5</span>) <span class="sc">+</span> <span class="fl">0.1</span>)</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>){</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>    ylimits <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">min</span>(bmodel_df_beta[, i]), <span class="fu">max</span>(bmodel_df_beta[, i]))</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>    <span class="co">#print(ylimits)</span></span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot</span>(bmodel_df_beta[, i], <span class="at">type=</span><span class="st">'l'</span>, <span class="at">col=</span>useColors[<span class="dv">1</span>], <span class="at">ylim=</span>ylimits, <span class="at">xaxt=</span><span class="st">'n'</span>, <span class="at">main=</span>df_beta_labels[i])</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>, <span class="at">labels=</span>track_names, <span class="at">at=</span><span class="fu">seq</span>(<span class="dv">1</span>, <span class="fu">nrow</span>(bmodel_df_beta)))</span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="st">'GATA3 tracks'</span>, <span class="at">side=</span><span class="dv">1</span>, <span class="at">outer=</span>T, <span class="at">cex=</span><span class="fl">1.2</span>,  <span class="at">line=</span><span class="dv">2</span>, <span class="at">adj=</span><span class="fl">0.4</span>)</span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="st">'coefficients'</span>, <span class="at">side=</span><span class="dv">2</span>, <span class="at">outer=</span>T, <span class="at">cex=</span><span class="fl">1.2</span>)</span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="fu">glue</span>(<span class="st">'{rownames(out)[bmodel_index]}: {used_categories}'</span>), <span class="at">side=</span><span class="dv">3</span>, <span class="at">outer=</span>T, <span class="at">cex=</span><span class="fl">1.3</span>, <span class="at">adj=</span><span class="dv">0</span>)</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mai=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>))</span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.new</span>()</span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">'right'</span>, <span class="at">legend=</span><span class="st">'TF ChIP-seq'</span>, <span class="at">col=</span>useColors[<span class="dv">1</span>], <span class="at">lty=</span><span class="dv">1</span>, <span class="at">bty=</span><span class="st">'n'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="impact-enformer-modelling-3_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>