<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.0.38">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Temi">
<meta name="dcterms.date" content="2022-09-28">

<title>Building models from Kawakami’s TF ChIP seq data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="kawakami-models_files/libs/clipboard/clipboard.min.js"></script>
<script src="kawakami-models_files/libs/quarto-html/quarto.js"></script>
<script src="kawakami-models_files/libs/quarto-html/popper.min.js"></script>
<script src="kawakami-models_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="kawakami-models_files/libs/quarto-html/anchor.min.js"></script>
<link href="kawakami-models_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="kawakami-models_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="kawakami-models_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="kawakami-models_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="kawakami-models_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Building models from Kawakami’s TF ChIP seq data</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Temi </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">September 28, 2022</p>
    </div>
  </div>
    
  </div>
  

</header>

<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list=</span><span class="fu">ls</span>())</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">'/projects/covid-ct/imlab/users/temi/projects/TFXcan/scripts/'</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glue)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: stats4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    anyDuplicated, append, as.data.frame, basename, cbind, colnames,
    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
    union, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: IRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'IRanges'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:glue':

    trim</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(R.utils)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: R.oo</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: R.methodsS3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>R.methodsS3 v1.8.2 (2022-06-13 22:00:14 UTC) successfully loaded. See ?R.methodsS3 for help.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>R.oo v1.25.0 (2022-06-12 02:20:02 UTC) successfully loaded. See ?R.oo for help.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'R.oo'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:R.methodsS3':

    throw</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:GenomicRanges':

    trim</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:IRanges':

    trim</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:glue':

    trim</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:methods':

    getClasses, getMethods</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    attach, detach, load, save</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>R.utils v2.12.0 (2022-06-28 03:20:05 UTC) successfully loaded. See ?R.utils for help.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'R.utils'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:utils':

    timestamp</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    cat, commandArgs, getOption, isOpen, nullfile, parse, warnings</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'data.table'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:GenomicRanges':

    shift</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:IRanges':

    shift</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:S4Vectors':

    first, second</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(glmnet)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Matrix</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'Matrix'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:S4Vectors':

    expand</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loaded glmnet 4.1-4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(doMC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: foreach</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: iterators</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: parallel</code></pre>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ROCR)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Matrix)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reshape2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'reshape2'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:data.table':

    dcast, melt</code></pre>
</div>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching packages
───────────────────────────────────────
tidyverse 1.3.2 ──</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>✔ ggplot2 3.3.6     ✔ purrr   0.3.4
✔ tibble  3.1.8     ✔ dplyr   1.0.9
✔ tidyr   1.2.0     ✔ stringr 1.4.0
✔ readr   2.1.2     ✔ forcats 0.5.1
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ purrr::accumulate() masks foreach::accumulate()
✖ dplyr::between()    masks data.table::between()
✖ dplyr::collapse()   masks IRanges::collapse()
✖ dplyr::combine()    masks BiocGenerics::combine()
✖ dplyr::desc()       masks IRanges::desc()
✖ tidyr::expand()     masks Matrix::expand(), S4Vectors::expand()
✖ tidyr::extract()    masks R.utils::extract()
✖ dplyr::filter()     masks stats::filter()
✖ dplyr::first()      masks data.table::first(), S4Vectors::first()
✖ dplyr::lag()        masks stats::lag()
✖ dplyr::last()       masks data.table::last()
✖ tidyr::pack()       masks Matrix::pack()
✖ ggplot2::Position() masks BiocGenerics::Position(), base::Position()
✖ purrr::reduce()     masks GenomicRanges::reduce(), IRanges::reduce()
✖ dplyr::rename()     masks S4Vectors::rename()
✖ dplyr::slice()      masks IRanges::slice()
✖ purrr::transpose()  masks data.table::transpose()
✖ tidyr::unpack()     masks Matrix::unpack()
✖ purrr::when()       masks foreach::when()</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>plots_dir <span class="ot">&lt;-</span> <span class="st">'../plots'</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>data_dir <span class="ot">&lt;-</span> <span class="st">'../enformer_predictions/train-test-val'</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>TF <span class="ot">&lt;-</span> <span class="st">'FOXA1'</span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>cell_line <span class="ot">&lt;-</span> <span class="st">'LuCaP'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For plotting, I need to load in the enformer annotations</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>enformer_annotations <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="st">'../info-files/enformer_tracks_annotated-resaved.txt'</span>)</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>enformer_annotations <span class="ot">&lt;-</span> enformer_annotations[<span class="sc">!</span><span class="fu">is.na</span>(enformer_annotations<span class="sc">$</span>assay), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>test_models <span class="ot">&lt;-</span> <span class="cf">function</span>(model, X_test_set, y_test_set){</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>    features <span class="ot">&lt;-</span> model<span class="sc">$</span>glmnet.fit<span class="sc">$</span>beta <span class="sc">|&gt;</span> <span class="fu">rownames</span>()</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>    X_test_set <span class="ot">&lt;-</span> X_test_set[, features]</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">assess.glmnet</span>(model, <span class="at">newx =</span> X_test_set, <span class="at">newy =</span> y_test_set) <span class="sc">|&gt;</span> <span class="fu">unlist</span>()</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Aggregating by the center</p>
<p>Loading the data -</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>dt_center <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="fu">glue</span>(<span class="st">'{data_dir}/aggByCenter_{TF}_{cell_line}.csv'</span>))</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>dt_center_0 <span class="ot">&lt;-</span> dt_center[dt_center<span class="sc">$</span>class <span class="sc">==</span> <span class="dv">0</span>, ]</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>dt_center_1 <span class="ot">&lt;-</span> dt_center[dt_center<span class="sc">$</span>class <span class="sc">==</span> <span class="dv">1</span>, ]</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>dt_center<span class="sc">$</span>class <span class="sc">|&gt;</span> <span class="fu">table</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   0    1 
1436 1325 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2022</span>)</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>train_indices_0 <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="at">x=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(dt_center_0), <span class="at">size=</span><span class="dv">1000</span>, <span class="at">replace=</span>F)</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>train_indices_1 <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="at">x=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(dt_center_1), <span class="at">size=</span><span class="dv">1000</span>, <span class="at">replace=</span>F)</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>train_data <span class="ot">&lt;-</span> <span class="fu">rbind</span>(dt_center_0[train_indices_0, ], dt_center_1[train_indices_1, ])</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>test_data <span class="ot">&lt;-</span> <span class="fu">rbind</span>(dt_center_0[<span class="sc">-</span>train_indices_0, ], dt_center_1[<span class="sc">-</span>train_indices_1, ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(train_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2000 5314</code></pre>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(test_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  761 5314</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>X_train <span class="ot">&lt;-</span> train_data[, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>)] <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>()</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>y_train <span class="ot">&lt;-</span> train_data[, <span class="dv">1</span>] <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>()</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>X_test <span class="ot">&lt;-</span> test_data[, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>)] <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>()</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>y_test <span class="ot">&lt;-</span> test_data[, <span class="dv">1</span>] <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Register a backend for parallelization</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co"># register a parallel backend </span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="co"># registerDoMC(cores = 4)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Fit an elastic net model</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># enet_center &lt;- glmnet::cv.glmnet(x=X_train[complete.cases(X_train),], y=y_train[complete.cases(X_train)], family = "binomial", </span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="co"># type.measure = "auc", alpha = 0.5, parallel=T, keep=T) #alpha: mixing term, lasso, 1-alpha ridge.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>build_models <span class="ot">&lt;-</span> <span class="cf">function</span>(aggMList, <span class="at">dt_dir=</span>data_dir){</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>    enet_models <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>    enet_performance <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>    enet_metrics <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(aggMList)){</span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>        aggM <span class="ot">&lt;-</span> aggMList[i]</span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a>        dt_center <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="fu">glue</span>(<span class="st">'{dt_dir}/{aggM}_{TF}_{cell_line}.csv'</span>))</span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a>        dt_center_0 <span class="ot">&lt;-</span> dt_center[dt_center<span class="sc">$</span>class <span class="sc">==</span> <span class="dv">0</span>, ]</span>
<span id="cb63-13"><a href="#cb63-13" aria-hidden="true" tabindex="-1"></a>        dt_center_1 <span class="ot">&lt;-</span> dt_center[dt_center<span class="sc">$</span>class <span class="sc">==</span> <span class="dv">1</span>, ]</span>
<span id="cb63-14"><a href="#cb63-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-15"><a href="#cb63-15" aria-hidden="true" tabindex="-1"></a>        <span class="fu">set.seed</span>(<span class="dv">2022</span>)</span>
<span id="cb63-16"><a href="#cb63-16" aria-hidden="true" tabindex="-1"></a>        train_indices_0 <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="at">x=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(dt_center_0), <span class="at">size=</span><span class="dv">1000</span>, <span class="at">replace=</span>F)</span>
<span id="cb63-17"><a href="#cb63-17" aria-hidden="true" tabindex="-1"></a>        train_indices_1 <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="at">x=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(dt_center_1), <span class="at">size=</span><span class="dv">1000</span>, <span class="at">replace=</span>F)</span>
<span id="cb63-18"><a href="#cb63-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-19"><a href="#cb63-19" aria-hidden="true" tabindex="-1"></a>        train_data <span class="ot">&lt;-</span> <span class="fu">rbind</span>(dt_center_0[train_indices_0, ], dt_center_1[train_indices_1, ])</span>
<span id="cb63-20"><a href="#cb63-20" aria-hidden="true" tabindex="-1"></a>        test_data <span class="ot">&lt;-</span> <span class="fu">rbind</span>(dt_center_0[<span class="sc">-</span>train_indices_0, ], dt_center_1[<span class="sc">-</span>train_indices_1, ])</span>
<span id="cb63-21"><a href="#cb63-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-22"><a href="#cb63-22" aria-hidden="true" tabindex="-1"></a>        X_train <span class="ot">&lt;-</span> train_data[, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>)] <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>()</span>
<span id="cb63-23"><a href="#cb63-23" aria-hidden="true" tabindex="-1"></a>        y_train <span class="ot">&lt;-</span> train_data[, <span class="dv">1</span>] <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>()</span>
<span id="cb63-24"><a href="#cb63-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-25"><a href="#cb63-25" aria-hidden="true" tabindex="-1"></a>        X_test <span class="ot">&lt;-</span> test_data[, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>)] <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>()</span>
<span id="cb63-26"><a href="#cb63-26" aria-hidden="true" tabindex="-1"></a>        y_test <span class="ot">&lt;-</span> test_data[, <span class="dv">1</span>] <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>()</span>
<span id="cb63-27"><a href="#cb63-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-28"><a href="#cb63-28" aria-hidden="true" tabindex="-1"></a>        enet_models[[i]] <span class="ot">&lt;-</span> glmnet<span class="sc">::</span><span class="fu">cv.glmnet</span>(<span class="at">x=</span>X_train[<span class="fu">complete.cases</span>(X_train),], <span class="at">y=</span>y_train[<span class="fu">complete.cases</span>(X_train)], <span class="at">family =</span> <span class="st">"binomial"</span>, <span class="at">type.measure =</span> <span class="st">"auc"</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">parallel=</span>T, <span class="at">keep=</span>T)</span>
<span id="cb63-29"><a href="#cb63-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-30"><a href="#cb63-30" aria-hidden="true" tabindex="-1"></a>        <span class="co"># performance</span></span>
<span id="cb63-31"><a href="#cb63-31" aria-hidden="true" tabindex="-1"></a>        pred_probs <span class="ot">&lt;-</span> <span class="fu">predict</span>(enet_models[[i]], X_test, <span class="at">type=</span><span class="st">'response'</span>)</span>
<span id="cb63-32"><a href="#cb63-32" aria-hidden="true" tabindex="-1"></a>        pred_instance <span class="ot">&lt;-</span> ROCR<span class="sc">::</span><span class="fu">prediction</span>(pred_probs, y_test)</span>
<span id="cb63-33"><a href="#cb63-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-34"><a href="#cb63-34" aria-hidden="true" tabindex="-1"></a>        test_auc_roc <span class="ot">&lt;-</span> ROCR<span class="sc">::</span><span class="fu">performance</span>(pred_instance, <span class="st">'tpr'</span>, <span class="st">'fpr'</span>)</span>
<span id="cb63-35"><a href="#cb63-35" aria-hidden="true" tabindex="-1"></a>        test_prec_rec <span class="ot">&lt;-</span> ROCR<span class="sc">::</span><span class="fu">performance</span>(pred_instance, <span class="st">"prec"</span>, <span class="st">"rec"</span>)</span>
<span id="cb63-36"><a href="#cb63-36" aria-hidden="true" tabindex="-1"></a>        enet_performance[[i]] <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">auc=</span>test_auc_roc, <span class="at">auprc=</span>test_prec_rec)</span>
<span id="cb63-37"><a href="#cb63-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-38"><a href="#cb63-38" aria-hidden="true" tabindex="-1"></a>        auc_value <span class="ot">&lt;-</span> glmnet<span class="sc">::</span><span class="fu">assess.glmnet</span>(enet_models[[i]], X_test, y_test)<span class="sc">$</span>auc[[<span class="dv">1</span>]]</span>
<span id="cb63-39"><a href="#cb63-39" aria-hidden="true" tabindex="-1"></a>        <span class="co">#prauc_value &lt;- ROCR::performance(pred_instance, "prec", "rec")</span></span>
<span id="cb63-40"><a href="#cb63-40" aria-hidden="true" tabindex="-1"></a>        enet_metrics[[i]] <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">auc_values=</span>auc_value)</span>
<span id="cb63-41"><a href="#cb63-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-42"><a href="#cb63-42" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb63-43"><a href="#cb63-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-44"><a href="#cb63-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(enet_models) <span class="ot">&lt;-</span> aggMList</span>
<span id="cb63-45"><a href="#cb63-45" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(enet_performance) <span class="ot">&lt;-</span> aggMList</span>
<span id="cb63-46"><a href="#cb63-46" aria-hidden="true" tabindex="-1"></a>    <span class="fu">names</span>(enet_metrics) <span class="ot">&lt;-</span> aggMList</span>
<span id="cb63-47"><a href="#cb63-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-48"><a href="#cb63-48" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">models=</span>enet_models, <span class="at">performance=</span>enet_performance, <span class="at">auc_values=</span>enet_metrics))  </span>
<span id="cb63-49"><a href="#cb63-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-50"><a href="#cb63-50" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>aggMethods <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'aggByCenter'</span>, <span class="st">'aggByMean'</span>, <span class="st">'aggByUpstream'</span>, <span class="st">'aggByDownstream'</span>, <span class="st">'aggByUpstreamDownstream'</span>)</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>enet_experiment <span class="ot">&lt;-</span> <span class="fu">build_models</span>(aggMethods)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: executing %dopar% sequentially: no parallel backend registered</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>colors_ <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'black'</span>, <span class="st">'red'</span>, <span class="st">'green'</span>, <span class="st">'blue'</span>, <span class="st">'grey'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">layout</span>(<span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">nrow=</span><span class="dv">1</span>, <span class="at">ncol=</span><span class="dv">2</span>))</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(enet_experiment<span class="sc">$</span>performance[[aggMethods[<span class="dv">1</span>]]]<span class="sc">$</span>auc, <span class="at">col=</span>colors_[<span class="dv">1</span>], <span class="at">main=</span><span class="st">'ROC curve | using all 5313 tracks from ENFORMER'</span>)</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(aggMethods)){</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot</span>(enet_experiment<span class="sc">$</span>performance[[aggMethods[i]]]<span class="sc">$</span>auc, <span class="at">col=</span>colors_[i], <span class="at">add=</span>T)</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">'center'</span>, <span class="at">legend=</span><span class="fu">paste0</span>(aggMethods, <span class="st">' ('</span>, <span class="fu">round</span>(enet_experiment<span class="sc">$</span>auc_values <span class="sc">|&gt;</span> <span class="fu">unlist</span>(), <span class="dv">3</span>) <span class="sc">|&gt;</span> <span class="fu">unname</span>(), <span class="st">')'</span>), <span class="at">col=</span>colors_, <span class="at">lty=</span><span class="dv">1</span>)</span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(enet_experiment<span class="sc">$</span>performance[[aggMethods[<span class="dv">1</span>]]]<span class="sc">$</span>auprc, <span class="at">col=</span>colors_[<span class="dv">1</span>], <span class="at">main=</span><span class="st">'Precision Recall curve | using all 5313 tracks from ENFORMER'</span>)</span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(aggMethods)){</span>
<span id="cb67-11"><a href="#cb67-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(aggMethods[i])</span>
<span id="cb67-12"><a href="#cb67-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot</span>(enet_experiment<span class="sc">$</span>performance[[aggMethods[i]]]<span class="sc">$</span>auprc, <span class="at">col=</span>colors_[i], <span class="at">add=</span>T)</span>
<span id="cb67-13"><a href="#cb67-13" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "aggByMean"
[1] "aggByUpstream"
[1] "aggByDownstream"
[1] "aggByUpstreamDownstream"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="at">x=</span><span class="fl">0.2</span>, <span class="at">y=</span><span class="fl">0.6</span>, <span class="at">legend=</span>aggMethods, <span class="at">col=</span>colors_, <span class="at">lty=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="kawakami-models_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">

</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot(enet_experiment$performance[[aggMethods[1]]]$auc, col=colors_[1])</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="co"># plot(enet_experiment$performance[[aggMethods[2]]]$auc, col=colors_[2], add=T)</span></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a><span class="co"># plot(enet_experiment$performance[[aggMethods[3]]]$auc, col=colors_[3], add=T)</span></span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a><span class="co"># plot(enet_experiment$performance[[aggMethods[4]]]$auc, col=colors_[4], add=T)</span></span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a><span class="co"># plot(enet_experiment$performance[[aggMethods[5]]]$auc, col=colors_[5], add=T)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Exploring the aggByCenter model</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>dimensions <span class="ot">&lt;-</span> enet_experiment<span class="sc">$</span>models<span class="sc">$</span>aggByCenter<span class="sc">$</span>glmnet.fit<span class="sc">$</span>beta<span class="sc">@</span>Dim</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>coef_mat <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">summary</span>(enet_experiment<span class="sc">$</span>models<span class="sc">$</span>aggByCenter<span class="sc">$</span>glmnet.fit<span class="sc">$</span>beta))</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(coef_mat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "i" "j" "x"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>temp_mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data=</span><span class="cn">NA</span>, <span class="at">nrow=</span>dimensions[<span class="dv">1</span>], <span class="at">ncol=</span>dimensions[<span class="dv">2</span>])</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(coef_mat)){</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>    temp_mat[coef_mat[i, <span class="st">'i'</span>], coef_mat[i, <span class="st">'j'</span>]] <span class="ot">&lt;-</span> coef_mat[i, <span class="st">'x'</span>]</span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a>temp_mat[<span class="fu">is.na</span>(temp_mat)] <span class="ot">&lt;-</span> <span class="dv">0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>min_error_index <span class="ot">&lt;-</span> enet_experiment<span class="sc">$</span>models<span class="sc">$</span>aggByCenter<span class="sc">$</span>index[<span class="st">'min'</span>, ]</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>one_sd_index <span class="ot">&lt;-</span> enet_experiment<span class="sc">$</span>models<span class="sc">$</span>aggByCenter<span class="sc">$</span>index[<span class="st">'1se'</span>, ]</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>df_beta <span class="ot">&lt;-</span> <span class="fu">cbind</span>(temp_mat[, min_error_index], temp_mat[, one_sd_index]) <span class="sc">|&gt;</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>df_beta_labels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'For lambda with minimum mean cv error'</span>, <span class="st">'For largest lambda with 1 s.e within the mininum error'</span>)</span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>motif_center <span class="ot">&lt;-</span> <span class="dv">448</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>df_beta<span class="sc">$</span>feature_names <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">'f'</span>, <span class="dv">0</span><span class="sc">:</span>(<span class="fu">nrow</span>(df_beta)<span class="sc">-</span><span class="dv">1</span>), <span class="at">sep=</span><span class="st">'_'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>feature_coefficients <span class="ot">&lt;-</span> <span class="fu">merge</span>(df_beta, enformer_annotations, <span class="at">by=</span><span class="st">'feature_names'</span>, <span class="at">all=</span>T)</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>feature_coefficients<span class="sc">$</span>sort_by <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">str_split</span>(feature_coefficients<span class="sc">$</span>feature_names, <span class="at">pattern =</span> <span class="st">'_'</span>), getElement, <span class="dv">2</span>) <span class="sc">|&gt;</span> <span class="fu">as.numeric</span>()</span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>feature_coefficients <span class="ot">&lt;-</span> feature_coefficients[<span class="fu">order</span>(feature_coefficients<span class="sc">$</span>sort_by), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(feature_coefficients[, <span class="st">'V1'</span>], <span class="at">type=</span><span class="st">'l'</span>, <span class="at">xlab=</span><span class="st">'tracks'</span>, <span class="at">xaxt=</span><span class="st">'n'</span>, <span class="at">ylab=</span><span class="st">'coefficients'</span>, <span class="at">main=</span><span class="st">'Coefficients across predicted ENFORMER tracks'</span>)</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>, <span class="at">at=</span><span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">5313</span>), <span class="at">col=</span><span class="fu">factor</span>(feature_coefficients<span class="sc">$</span>mark))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="kawakami-models_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co"># #abline(v=motif_center, col='red')</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a><span class="co"># mtext('tracks', side=1, outer=T, cex=1.2)</span></span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a><span class="co"># mtext('coefficients', side=2, outer=T, cex=1.2)</span></span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a><span class="co"># mtext('Distribution of coefficients across tracks; aggByMean; except GATA3', side=3, outer=T, cex=1.3)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>useColors <span class="ot">&lt;-</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="at">n=</span><span class="dv">8</span>, <span class="at">name=</span><span class="st">'Dark2'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>feature_coefficients_split <span class="ot">&lt;-</span> <span class="fu">split</span>(feature_coefficients, feature_coefficients<span class="sc">$</span>assay)</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>ylimits <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">min</span>(feature_coefficients[, <span class="st">'V1'</span>]), <span class="fu">max</span>(feature_coefficients[, <span class="st">'V1'</span>]))</span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a><span class="co"># |&gt; row.names() |&gt; as.numeric()</span></span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x=</span>feature_coefficients_split[[<span class="dv">1</span>]]<span class="sc">$</span>sort_by , <span class="at">y=</span>feature_coefficients_split[[<span class="dv">1</span>]][, <span class="st">'V1'</span>], <span class="at">type=</span><span class="st">'l'</span>, <span class="at">col=</span>useColors[<span class="dv">1</span>], <span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">1</span>, <span class="fu">nrow</span>(feature_coefficients)), <span class="at">ylim=</span>ylimits, <span class="at">xaxt=</span><span class="st">'n'</span>, <span class="at">main=</span><span class="st">'Coefficients across 5313 ENFORMER tracks'</span>, <span class="at">xlab=</span><span class="st">'ENFORMER tracks'</span>, <span class="at">ylab=</span><span class="st">'coefficients'</span>)</span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="fu">length</span>(feature_coefficients_split)){</span>
<span id="cb80-8"><a href="#cb80-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">lines</span>(<span class="at">x=</span>feature_coefficients_split[[j]]<span class="sc">$</span>sort_by, <span class="at">y=</span>feature_coefficients_split[[j]][, <span class="st">'V1'</span>], <span class="at">col=</span>useColors[j])</span>
<span id="cb80-9"><a href="#cb80-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb80-10"><a href="#cb80-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-11"><a href="#cb80-11" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>(<span class="dv">1</span>, <span class="at">at=</span><span class="fu">seq</span>(<span class="dv">1</span>, <span class="fu">nrow</span>(feature_coefficients)))</span>
<span id="cb80-12"><a href="#cb80-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-13"><a href="#cb80-13" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">'bottomleft'</span>, <span class="at">legend=</span><span class="fu">names</span>(feature_coefficients_split), <span class="at">col=</span>useColors[<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(feature_coefficients_split)], <span class="at">lty=</span><span class="dv">1</span>, <span class="at">bty=</span><span class="st">'n'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="kawakami-models_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>nfcsplit <span class="ot">&lt;-</span> <span class="fu">names</span>(feature_coefficients_split)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>usage_metric <span class="ot">&lt;-</span> <span class="fu">lapply</span>(feature_coefficients_split, <span class="cf">function</span>(each_cat){</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>    usage <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(each_cat<span class="sc">$</span>V1 <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>)</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>( <span class="fu">paste</span>(<span class="fu">sum</span>(usage), <span class="st">'of'</span>, <span class="fu">length</span>(usage), <span class="at">sep=</span><span class="st">' '</span>))</span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>})  <span class="sc">|&gt;</span> <span class="fu">unlist</span>()</span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>usage_metric</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        ATAC-seq  CAGE experiment        DNase-seq Histone ChIP-seq 
       "3 of 10"       "6 of 638"      "12 of 674"     "56 of 2000" 
     TF ChIP-seq 
    "42 of 1990" </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>usage_heights <span class="ot">&lt;-</span> <span class="fu">lapply</span>(feature_coefficients_split, <span class="cf">function</span>(each_cat){</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>    usage <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(each_cat<span class="sc">$</span>V1 <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>)</span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>((<span class="fu">sum</span>(usage) <span class="sc">/</span> <span class="fu">length</span>(usage)) <span class="sc">*</span> <span class="dv">100</span>)</span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">|&gt;</span> <span class="fu">unlist</span>()</span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a>bplot <span class="ot">&lt;-</span> <span class="fu">barplot</span>(usage_heights, <span class="at">main=</span><span class="st">'Percent usage of tracks per category'</span>, <span class="at">ylab=</span><span class="st">'% of tracks with non-zero coefficients'</span>, <span class="at">xlab=</span><span class="st">'category'</span>, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">35</span>), <span class="at">col=</span><span class="st">"#69b3a2"</span>)</span>
<span id="cb84-7"><a href="#cb84-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-8"><a href="#cb84-8" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x=</span>bplot, <span class="at">y=</span>usage_heights <span class="sc">+</span> <span class="dv">1</span>, <span class="at">pos =</span> <span class="dv">3</span>, <span class="at">label =</span> usage_metric, <span class="at">cex =</span> <span class="dv">1</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="kawakami-models_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>coef_details <span class="ot">&lt;-</span> feature_coefficients[<span class="fu">abs</span>(feature_coefficients<span class="sc">$</span>V1) <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="fu">c</span>(<span class="st">'V1'</span>, <span class="st">'assay'</span>, <span class="st">'detail'</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>coef_details[<span class="fu">grepl</span>(<span class="st">'FOXA1'</span>, coef_details<span class="sc">$</span>detail), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              V1       assay
5263 0.008505939 TF ChIP-seq
24   2.350618085 TF ChIP-seq
953  0.326976919 TF ChIP-seq
                                                          detail
5263 FOXA1:T47D treated with 0.02% dimethyl sulfoxide for 1 hour
24       FOXA1:A549 treated with 100 nM dexamethasone for 1 hour
953                                                  FOXA1:MCF-7</code></pre>
</div>
</div>
<section id="doing-all-different-combinations" class="level1">
<h1>Doing all different combinations</h1>
<div class="cell">

</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.1.3 (2022-03-10)
Platform: x86_64-conda-linux-gnu (64-bit)
Running under: SUSE Linux Enterprise Server 15 SP1

Matrix products: default
BLAS/LAPACK: /lus/swift/home/temi/miniconda3/envs/r-env/lib/libopenblasp-r0.3.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] RColorBrewer_1.1-3   forcats_0.5.1        stringr_1.4.0       
 [4] dplyr_1.0.9          purrr_0.3.4          readr_2.1.2         
 [7] tidyr_1.2.0          tibble_3.1.8         ggplot2_3.3.6       
[10] tidyverse_1.3.2      reshape2_1.4.4       ROCR_1.0-11         
[13] doMC_1.3.8           iterators_1.0.14     foreach_1.5.2       
[16] glmnet_4.1-4         Matrix_1.4-1         data.table_1.14.2   
[19] R.utils_2.12.0       R.oo_1.25.0          R.methodsS3_1.8.2   
[22] reticulate_1.25      GenomicRanges_1.46.1 GenomeInfoDb_1.30.1 
[25] IRanges_2.28.0       S4Vectors_0.32.4     BiocGenerics_0.40.0 
[28] glue_1.6.2          

loaded via a namespace (and not attached):
 [1] httr_1.4.4             jsonlite_1.8.0         splines_4.1.3         
 [4] modelr_0.1.8           assertthat_0.2.1       GenomeInfoDbData_1.2.7
 [7] googlesheets4_1.0.1    cellranger_1.1.0       yaml_2.3.5            
[10] pillar_1.8.0           backports_1.4.1        lattice_0.20-45       
[13] digest_0.6.29          XVector_0.34.0         rvest_1.0.2           
[16] colorspace_2.0-3       htmltools_0.5.3        plyr_1.8.7            
[19] pkgconfig_2.0.3        broom_1.0.0            haven_2.5.0           
[22] zlibbioc_1.40.0        scales_1.2.0           tzdb_0.3.0            
[25] googledrive_2.0.0      generics_0.1.3         ellipsis_0.3.2        
[28] withr_2.5.0            cli_3.3.0              crayon_1.5.1          
[31] readxl_1.4.0           survival_3.4-0         magrittr_2.0.3        
[34] evaluate_0.16          fs_1.5.2               fansi_1.0.3           
[37] xml2_1.3.3             tools_4.1.3            hms_1.1.1             
[40] gargle_1.2.0           lifecycle_1.0.1        reprex_2.0.2          
[43] munsell_0.5.0          compiler_4.1.3         rlang_1.0.4           
[46] grid_4.1.3             RCurl_1.98-1.8         htmlwidgets_1.5.4     
[49] bitops_1.0-7           rmarkdown_2.15         gtable_0.3.0          
[52] codetools_0.2-18       DBI_1.1.3              R6_2.5.1              
[55] lubridate_1.8.0        knitr_1.39             fastmap_1.1.0         
[58] utf8_1.2.2             shape_1.4.6            stringi_1.7.8         
[61] Rcpp_1.0.9             vctrs_0.4.1            png_0.1-7             
[64] dbplyr_2.2.1           tidyselect_1.1.2       xfun_0.32             </code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>