
```{r import libraries}
setwd('/projects/covid-ct/imlab/users/temi/projects/TFXcan/scripts/')

rm(list=ls())

library(glue)
library(GenomicRanges)
library(reticulate)
library(R.utils)
library(data.table)
library(tidyverse)
```

```{r}
# pick a transcription factor
TF <- 'FOXA1' #
cell_line <- 'LuCaP' # a prostrate cancer cell line
foxa1_motif <- 'foxa1.lncap.motif'

# where is the TF chip-seq data, as well as other directories
work_dir <- '/projects/covid-ct/imlab/users/temi/projects/TFXcan'
common_files_dir <- glue('{work_dir}/data/common-files/{TF}')

#kawakami_dir <- glue('{work_dir}/data/kawakami-human')
kawakami_impact_dir <- glue('{work_dir}/data/kawakami-human')
kawakami_impact_split_dir <- glue('{kawakami_impact_dir}/kawakami-impact-split')

freedman_dir <- glue('{work_dir}/data/freedman')
freedman_impact_dir <- glue('{work_dir}/data/freedman')
freedman_impact_split_dir <- glue('{freedman_dir}/freedman-impact-split')
```

# read in the freedman peak info
```{r}

```
