---
title: "Exploring an enhancer database"
author: "Temi"
date: 'Wed June 28 2023'
format: 
  pdf: 
    toc: true
    number-sections: true
    code-line-numbers: true
---

```{r}
rm(list=ls())
expr_dir <- '/lus/grand/projects/TFXcan/imlab/users/temi/projects/TFXcan/experiments'
```

```{r}
library(devtools)
library(glue)
library(GenomicRanges)
library(reticulate)
library(R.utils)
library(data.table)
library(tidyverse)
library(annotatr)
library(AnnotationHub)
library(biomaRt)
library(hexbin)
library(RColorBrewer)
library(preprocessCore)
library(grid)
library(gridExtra)
```

### After predictions

```{r}
#setwd('/grand/TFXcan/imlab/users/temi/projects/TFXcan/scripts/')
project_dir <- '/lus/grand/projects/TFXcan/imlab/users/temi/projects/TFXcan'
enhancers_dir <- '/lus/grand/projects/TFXcan/imlab/data/enhancers'

# for qqplots:
#devtools::source_gist('https://gist.github.com/TemiPete/250d9922b9516691f83bb1fd999a3ccc')

# for manhattan plots
devtools::source_gist('https://gist.github.com/TemiPete/250d9922b9516691f83bb1fd999a3ccc')
devtools::source_gist('https://gist.github.com/hakyim/38431b74c6c0bf90c12f')
devtools::source_gist('https://gist.github.com/hakyim/5d2251ea1a86009499e4ffdf47fe2735')
```


```{r}
enh_genes <- data.table::fread(glue('{enhancers_dir}/ENdb_target_gene.txt'))
all_enh <- data.table::fread(glue('{enhancers_dir}/ENdb_enhancer.txt')) %>% 
    dplyr::filter(Species == 'Human') %>% dplyr::select(Enhancer_id, Enhancer_symbol, Reference_genome, Chromosome, Start_position, End_position, Species, Biosample_name, Target_gene, target_gene_other_name, Disease, TF_name, TF_other_name, SNP_id, SNP_position, Distance_from_TSS, n_Distance_from_TSS)
dim(enh_genes) ; enh_genes[1:5, ] ; dim(all_enh) ; all_enh[1:5, ]
```

```{r}
enh_genes_maps <- dplyr::left_join(enh_genes, all_enh, by=c('E_ID' = 'Enhancer_id'))
dim(enh_genes_maps) ; enh_genes_maps[1:5, 1:5]
```


```{r}
enh_genes_maps %>% dplyr::group_by(target_gene) %>% dplyr::summarise(cnt = n()) %>% dplyr::arrange(desc(cnt)) #%>% dplyr::pull(cnt) %>% hist(main='number of enhancers per gene', xlab='number of enhancers')
```

```{r}
enh_genes_maps %>% 
    dplyr::filter(!is.na(Start_position | End_position)) %>% 
    dplyr::group_by(TF_name) %>% 
    dplyr::summarise(cnt = n()) %>% 
    dplyr::arrange(desc(cnt)) %>% 
    dplyr::pull(cnt) %>% hist(main='number of enhancers per gene', xlab='number of enhancers')
```

```{r}
enh_genes_maps %>% dplyr::filter(!is.na(Start_position | End_position))
```


```{r}
enh_genes_maps %>% 
    dplyr::filter(!is.na(Start_position | End_position)) %>%
    dplyr::select(Start_position, End_position) %>% 
    dplyr::mutate(enh_dist = End_position - Start_position) %>% dplyr::pull(enh_dist) %>% hist(breaks = 80)
```